{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identifying relationships between annotated omics data in NMDC\n",
    "\n",
    "In this notebook, we’ll look at how different types of omics data can be connected using popular annotation vocabularies. We’ll focus on exploring biomolecules and KEGG pathways from a group of samples that have both metagenomic and metatranscriptomic data, all available in the NMDC Data Portal. By working through these examples, you’ll see how to bring these data types together for integrated analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "## First need to install locally the nmdc_api_utilities\n",
    "%pip install nmdc_api_utilities\n",
    "%pip install dotenv\n",
    "%pip install plotly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up environment variables\n",
    "**You can disregard this section of code unless you have interest in testing this on the development API.**\n",
    "\n",
    "Using pythons python-dotenv package, load the environment variables from the system. This chunk of code is used in the Github CI/CD pipelines to test our development API. The variable can be used when creating `nmdc_api_utilities` objects. If you do not have the environment variable `ENV` set in a .env file, this code with default to the variable \"prod\" - which tells nmdc_api_utilities to use the production API URL."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up environment variables\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "load_dotenv()\n",
    "# load the environment variable ENV. If it does not exist, default to \"prod\"\n",
    "ENV = os.environ.get(\"ENV\", \"prod\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import re, sys, io, gzip, json, time, math, hashlib\n",
    "import pandas as pd\n",
    "import plotly.express as px, plotly.graph_objects as go\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "from typing import Dict, List, Optional, Tuple\n",
    "from io import StringIO\n",
    "import nmdc_api_utilities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identifying relationships between annotated omics data in NMDC\n",
    "\n",
    "This notebook is an example of how different omics data types may be linked via commonly used annotation vocabularies and investigated together. In this notebook we explore KEGG pathways and taxonomy lineage identified in a set of samples that have processed metagenomics and metatranscriptomics data available in the NMDC Data Portal."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Retrieve data from the NMDC database using API endpoints\n",
    "\n",
    "### Choose data to retrieve\n",
    "\n",
    "The [NMDC Data Portal](https://data.microbiomedata.org/) is a powerful resource where you can search for samples by all kinds of criteria. For this example, we’ll focus on finding samples that have both metagenomics and metatranscriptomics data.  \n",
    "To do this, just use the Data Type filters (the “upset plot” below the interactive map) to quickly spot the right samples. In our case, these filters lead us to samples from the study [“Jeff Blanchard’s Harvard forest soil project”](https://data.microbiomedata.org/details/study/nmdc:sty-11-8ws97026).\n",
    "\n",
    "### Retrieve and filter data for Harvard forest soil study\n",
    "\n",
    "The study page linked above has the NMDC study identifier in the URL: `nmdc:sty-11-8ws97026`. We will use the [nmdc_api_utilities](https://microbiomedata.github.io/nmdc_api_utilities/) package to access the data_objects/study endpoint from the [NMDC Runtime API](https://api.microbiomedata.org/docs) to retrieve all records that represent data. This includes URLs for downloading raw data files (e.g. FASTQ) as well as processed data results output by the NMDC workflows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>name</th>\n",
       "      <th>file_size_bytes</th>\n",
       "      <th>md5_checksum</th>\n",
       "      <th>data_object_type</th>\n",
       "      <th>was_generated_by</th>\n",
       "      <th>url</th>\n",
       "      <th>description</th>\n",
       "      <th>data_category</th>\n",
       "      <th>biosample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nmdc:dobj-11-0ebr0z48</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>11862.7.224581.ACGGAAC-TGTTCCG.fastq.gz</td>\n",
       "      <td>34934525960</td>\n",
       "      <td>952d3a6780ac34a03b112884ec10cdda</td>\n",
       "      <td>Metagenome Raw Reads</td>\n",
       "      <td>nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>Metagenome Raw Reads for nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>instrument_data</td>\n",
       "      <td>nmdc:bsm-11-622k6044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nmdc:dobj-11-s5chx118</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>nmdc_wfrqc-11-840m6332.1_filtered.fastq.gz</td>\n",
       "      <td>31299212828</td>\n",
       "      <td>440a9bfb671ab0fea36359c715bb5f4e</td>\n",
       "      <td>Filtered Sequencing Reads</td>\n",
       "      <td>nmdc:wfrqc-11-840m6332.1</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>Reads QC for nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-622k6044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nmdc:dobj-11-p7q8bb28</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>nmdc_wfrqc-11-840m6332.1_filterStats.txt</td>\n",
       "      <td>801</td>\n",
       "      <td>b82b37fb53861fa34a06ca201068d727</td>\n",
       "      <td>QC Statistics</td>\n",
       "      <td>nmdc:wfrqc-11-840m6332.1</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>Reads QC summary for nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-622k6044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3346</th>\n",
       "      <td>nmdc:dobj-11-8pbtm711</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...</td>\n",
       "      <td>12668</td>\n",
       "      <td>da7f3a98506322255816e3c2507a11e9</td>\n",
       "      <td>Configuration toml</td>\n",
       "      <td>nmdc:wfmb-11-g1vtvp02.1</td>\n",
       "      <td>https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...</td>\n",
       "      <td>CoreMS parameters used for Lipidomics workflow.</td>\n",
       "      <td>workflow_parameter_data</td>\n",
       "      <td>nmdc:bsm-11-y52p4f86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3347</th>\n",
       "      <td>nmdc:dobj-11-869nnh02</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...</td>\n",
       "      <td>853274930</td>\n",
       "      <td>946f895d2a641e9dac6cd927715df3b8</td>\n",
       "      <td>LC-MS Lipidomics Processed Data</td>\n",
       "      <td>nmdc:wfmb-11-g1vtvp02.1</td>\n",
       "      <td>https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...</td>\n",
       "      <td>CoreMS hdf5 file representing a lipidomics dat...</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-y52p4f86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348</th>\n",
       "      <td>nmdc:dobj-11-zrzcp186</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...</td>\n",
       "      <td>31914586</td>\n",
       "      <td>5ec14469476f651fceaf2ec5068ad140</td>\n",
       "      <td>LC-MS Lipidomics Results</td>\n",
       "      <td>nmdc:wfmb-11-g1vtvp02.1</td>\n",
       "      <td>https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...</td>\n",
       "      <td>Lipid annotations as a result of a lipidomics ...</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-y52p4f86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3349 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id             type  \\\n",
       "0     nmdc:dobj-11-0ebr0z48  nmdc:DataObject   \n",
       "1     nmdc:dobj-11-s5chx118  nmdc:DataObject   \n",
       "2     nmdc:dobj-11-p7q8bb28  nmdc:DataObject   \n",
       "...                     ...              ...   \n",
       "3346  nmdc:dobj-11-8pbtm711  nmdc:DataObject   \n",
       "3347  nmdc:dobj-11-869nnh02  nmdc:DataObject   \n",
       "3348  nmdc:dobj-11-zrzcp186  nmdc:DataObject   \n",
       "\n",
       "                                                   name  file_size_bytes  \\\n",
       "0               11862.7.224581.ACGGAAC-TGTTCCG.fastq.gz      34934525960   \n",
       "1            nmdc_wfrqc-11-840m6332.1_filtered.fastq.gz      31299212828   \n",
       "2              nmdc_wfrqc-11-840m6332.1_filterStats.txt              801   \n",
       "...                                                 ...              ...   \n",
       "3346  Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...            12668   \n",
       "3347  Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...        853274930   \n",
       "3348  Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...         31914586   \n",
       "\n",
       "                          md5_checksum                 data_object_type  \\\n",
       "0     952d3a6780ac34a03b112884ec10cdda             Metagenome Raw Reads   \n",
       "1     440a9bfb671ab0fea36359c715bb5f4e        Filtered Sequencing Reads   \n",
       "2     b82b37fb53861fa34a06ca201068d727                    QC Statistics   \n",
       "...                                ...                              ...   \n",
       "3346  da7f3a98506322255816e3c2507a11e9               Configuration toml   \n",
       "3347  946f895d2a641e9dac6cd927715df3b8  LC-MS Lipidomics Processed Data   \n",
       "3348  5ec14469476f651fceaf2ec5068ad140         LC-MS Lipidomics Results   \n",
       "\n",
       "              was_generated_by  \\\n",
       "0        nmdc:dgns-11-0g2bvk46   \n",
       "1     nmdc:wfrqc-11-840m6332.1   \n",
       "2     nmdc:wfrqc-11-840m6332.1   \n",
       "...                        ...   \n",
       "3346   nmdc:wfmb-11-g1vtvp02.1   \n",
       "3347   nmdc:wfmb-11-g1vtvp02.1   \n",
       "3348   nmdc:wfmb-11-g1vtvp02.1   \n",
       "\n",
       "                                                    url  \\\n",
       "0     https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "1     https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "2     https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "...                                                 ...   \n",
       "3346  https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...   \n",
       "3347  https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...   \n",
       "3348  https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...   \n",
       "\n",
       "                                            description  \\\n",
       "0        Metagenome Raw Reads for nmdc:dgns-11-0g2bvk46   \n",
       "1                    Reads QC for nmdc:dgns-11-0g2bvk46   \n",
       "2            Reads QC summary for nmdc:dgns-11-0g2bvk46   \n",
       "...                                                 ...   \n",
       "3346    CoreMS parameters used for Lipidomics workflow.   \n",
       "3347  CoreMS hdf5 file representing a lipidomics dat...   \n",
       "3348  Lipid annotations as a result of a lipidomics ...   \n",
       "\n",
       "                data_category          biosample_id  \n",
       "0             instrument_data  nmdc:bsm-11-622k6044  \n",
       "1              processed_data  nmdc:bsm-11-622k6044  \n",
       "2              processed_data  nmdc:bsm-11-622k6044  \n",
       "...                       ...                   ...  \n",
       "3346  workflow_parameter_data  nmdc:bsm-11-y52p4f86  \n",
       "3347           processed_data  nmdc:bsm-11-y52p4f86  \n",
       "3348           processed_data  nmdc:bsm-11-y52p4f86  \n",
       "\n",
       "[3349 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.set_option(\"display.max_rows\", 6)\n",
    "from nmdc_api_utilities.data_object_search import DataObjectSearch\n",
    "do_client = DataObjectSearch(env=ENV)\n",
    "#get all data objects associated with this study id\n",
    "data = do_client.get_data_objects_for_studies(study_id='nmdc:sty-11-8ws97026')\n",
    "data = pd.DataFrame(data)\n",
    "\n",
    "#reformat data into dataframe\n",
    "data_objects=[]\n",
    "for index, row in data.iterrows():\n",
    "    bio_id = row['biosample_id']\n",
    "    row_out = pd.json_normalize(row['data_objects'])\n",
    "    row_out['biosample_id'] = bio_id\n",
    "    data_objects.append(row_out)\n",
    "\n",
    "data_objects = pd.concat(data_objects).reset_index(drop=True)\n",
    "display(data_objects)\n",
    "\n",
    "del data, index, row, row_out, bio_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One way of further identifying a NMDC `DataObject` record is by looking at its slot `data_object_type` (https://microbiomedata.github.io/nmdc-schema/data_object_type/), which contains a value from `FileTypeEnum` (https://microbiomedata.github.io/nmdc-schema/FileTypeEnum/). \n",
    "\n",
    "We want to look at the processed data results for our three omics types of interest in this notebook. Specifically, we want the files containing KEGG Orthology, Taxonomy Lineage, and  Expression data. These expression data is available in file for metatranscriptomics data (in \"Metatranscriptome Expression\" data object) and annotation data are found together in the metagenomics and metatranscriptomics ( in 'Annotation KEGG Orthology' and 'Gene Phylogeny tsv') results files.\n",
    "\n",
    "We filter for results files with the following `data_object_type` values:\n",
    "\n",
    "| Value | Description |\n",
    "|:-----:|:-----------:|\n",
    "|Gene Phylogeny tsv|Tab-delimited file of gene phylogeny|\n",
    "|Annotation KEGG Orthology|Tab delimited file for KO annotation|\n",
    "|Metatranscriptome Expression|Read count table output|\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>data_object_type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Annotation KEGG Orthology</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Phylogeny tsv</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Metatranscriptome Expression</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              count\n",
       "data_object_type                   \n",
       "Annotation KEGG Orthology        50\n",
       "Gene Phylogeny tsv               50\n",
       "Metatranscriptome Expression     31"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to biosamples that have KO annotations, Lineage, and expression results\n",
    "data_objects = data_objects.groupby('biosample_id').filter(lambda x: all(data_type in x['data_object_type'].values for data_type in [\n",
    "        \"Annotation KEGG Orthology\",\n",
    "        \"Metatranscriptome Expression\",\n",
    "        \"Gene Phylogeny tsv\"\n",
    "    ])).reset_index(drop=True)\n",
    "\n",
    "# Filter to the desired results file types\n",
    "results_by_biosample = data_objects[data_objects['data_object_type'].isin([\n",
    "    \"Annotation KEGG Orthology\",\n",
    "    \"Metatranscriptome Expression\", \"Gene Phylogeny tsv\"\n",
    "])][['biosample_id', 'id','data_object_type', 'url']].rename(columns={'id':'output_id'}).reset_index(drop=True)\n",
    "\n",
    "# Is there one output for each data object type for each biosample?\n",
    "results_by_biosample['data_object_type'].value_counts().to_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select data has both metaT and metaG analysis result\n",
    "Notice that Each biosample is expected to have two lineage TSV files and two KO results. Therefore, the total count should be double the expression data count (31 × 2 = 62).  However, we only observed 50. This indicates that some biosamples contain only one type of data.  \n",
    "\n",
    "To address this, we need to filter out those incomplete biosamples and retain only those with both metaG and metaT data for the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  count  count\n",
      "0  both_metaT_and_metaG     19\n",
      "1            metaT_only     12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>data_object_type</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>MetaG Annotation KEGG Orthology</th>\n",
       "      <th>MetaG Gene Phylogeny tsv</th>\n",
       "      <th>MetaT Annotation KEGG Orthology</th>\n",
       "      <th>MetaT Gene Phylogeny tsv</th>\n",
       "      <th>Metatranscriptome Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nmdc:bsm-11-042nd237</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nmdc:bsm-11-3p2mnz08</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nmdc:bsm-11-65a4xw75</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nmdc:bsm-11-8rn9bm20</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nmdc:bsm-11-cwey4y35</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nmdc:bsm-11-frgt4x11</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "data_object_type          biosample_id  \\\n",
       "0                 nmdc:bsm-11-042nd237   \n",
       "1                 nmdc:bsm-11-3p2mnz08   \n",
       "2                 nmdc:bsm-11-65a4xw75   \n",
       "..                                 ...   \n",
       "7                 nmdc:bsm-11-8rn9bm20   \n",
       "8                 nmdc:bsm-11-cwey4y35   \n",
       "9                 nmdc:bsm-11-frgt4x11   \n",
       "\n",
       "data_object_type                    MetaG Annotation KEGG Orthology  \\\n",
       "0                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "1                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "2                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "..                                                              ...   \n",
       "7                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "8                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "9                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "\n",
       "data_object_type                           MetaG Gene Phylogeny tsv  \\\n",
       "0                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "1                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "2                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "..                                                              ...   \n",
       "7                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "8                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "9                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "\n",
       "data_object_type                    MetaT Annotation KEGG Orthology  \\\n",
       "0                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "1                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "2                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "..                                                              ...   \n",
       "7                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "8                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "9                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "\n",
       "data_object_type                           MetaT Gene Phylogeny tsv  \\\n",
       "0                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "1                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "2                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "..                                                              ...   \n",
       "7                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "8                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "9                 https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "\n",
       "data_object_type                       Metatranscriptome Expression  \n",
       "0                 https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "1                 https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "2                 https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "..                                                              ...  \n",
       "7                 https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "8                 https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "9                 https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "\n",
       "[10 rows x 6 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_by_biosample[\"_has_metaT\"] = results_by_biosample[\"url\"].str.contains(r\"nmdc:wfmt\", case=False, na=False)\n",
    "results_by_biosample[\"_has_metaG\"] = results_by_biosample[\"url\"].str.contains(r\"nmdc:wfmg\", case=False, na=False)\n",
    "\n",
    "mask_annotation = results_by_biosample[\"data_object_type\"] == \"Annotation KEGG Orthology\"\n",
    "mask_lineage = results_by_biosample[\"data_object_type\"] == \"Gene Phylogeny tsv\"\n",
    "\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaT\"] & mask_annotation, \"data_object_type\"] = \"MetaT Annotation KEGG Orthology\"\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaG\"] & mask_annotation, \"data_object_type\"] = \"MetaG Annotation KEGG Orthology\"\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaT\"] & mask_lineage, \"data_object_type\"] = \"MetaT Gene Phylogeny tsv\"\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaG\"] & mask_lineage, \"data_object_type\"] = \"MetaG Gene Phylogeny tsv\"\n",
    "\n",
    "# Per-biosample aggregates\n",
    "agg = results_by_biosample.groupby(results_by_biosample[\"biosample_id\"]).agg(\n",
    "    rows_per_biosample=(\"biosample_id\", \"size\"),\n",
    "    has_metaT=(\"_has_metaT\", \"any\"),\n",
    "    has_metaG=(\"_has_metaG\", \"any\"),\n",
    "    metaT_row_count=(\"_has_metaT\", \"sum\"),\n",
    "    metaG_row_count=(\"_has_metaG\", \"sum\"),\n",
    ").reset_index()\n",
    "# Classify\n",
    "def classify(row):\n",
    "    if row[\"has_metaT\"] and row[\"has_metaG\"]:\n",
    "        return \"both_metaT_and_metaG\"\n",
    "    elif row[\"has_metaT\"] and not row[\"has_metaG\"]:\n",
    "        return \"metaT_only\"\n",
    "    elif row[\"has_metaG\"] and not row[\"has_metaT\"]:\n",
    "        return \"metaG_only\"\n",
    "    else:\n",
    "        return \"neither\"\n",
    "agg[\"category\"] = agg.apply(classify, axis=1)\n",
    "\n",
    "\n",
    "\n",
    "counts = (\n",
    "    agg[\"category\"]\n",
    "    .value_counts()\n",
    "    .reset_index()\n",
    "    .rename(columns={\"index\": \"category\", \"category\": \"count\"})\n",
    ")\n",
    "\n",
    "print(counts)\n",
    "\n",
    "\n",
    "# Filter to keep only biosamples with both\n",
    "keep_ids = set(agg.loc[agg[\"category\"] == \"both_metaT_and_metaG\", \"biosample_id\"])\n",
    "drop_ids = set(agg.loc[agg[\"category\"] != \"both_metaT_and_metaG\", \"biosample_id\"])\n",
    "filtered_results_by_biosample = results_by_biosample[results_by_biosample[\"biosample_id\"].isin(keep_ids)].drop(columns=[\"_has_metaT\", \"_has_metaG\"])\n",
    "wide = filtered_results_by_biosample.pivot(index=\"biosample_id\", columns=\"data_object_type\", values=\"url\").reset_index()\n",
    "wide.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download selected results files\n",
    " \n",
    "Now we can use the `url` slot from the filtered `DataObject` records to read in all of the files containing the annotations of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ingest into per-sample datastore\n",
    "ko_columns = [\"gene_id\", \"img_ko_flag\", \"ko_term\", \"percent_identity\",\n",
    "              \"query_start\", \"query_end\", \"subj_start\", \"subj_end\",\n",
    "              \"evalue\", \"bit_score\", \"align_length\"]\n",
    "lineage_columns = [\"gene_id\", \"homolog_gene_oid\", \"homolog_taxon_oid\", \"percent_identity\",\"lineage\"]\n",
    "\n",
    "# Function to read TSV files\n",
    "def read_tsv_from_url(url:str, column_names:None):\n",
    "    response = requests.get(url)\n",
    "    if column_names is None:\n",
    "        data = pd.read_csv(StringIO(response.text), sep='\\t')\n",
    "    else:\n",
    "        data = pd.read_csv(StringIO(response.text), sep='\\t', header=None, names=column_names)\n",
    "    return data\n",
    "\n",
    "def load_sample_row(row: pd.Series) -> Dict[str, Optional[pd.DataFrame]]:\n",
    "    col_map = {\n",
    "        \"MetaG Annotation KEGG Orthology\": (\"metag_ko_results\", ko_columns),\n",
    "        \"MetaT Annotation KEGG Orthology\": (\"metat_ko_results\", ko_columns),\n",
    "        \"MetaG Gene Phylogeny tsv\": (\"metag_lineage_results\", lineage_columns),\n",
    "        \"MetaT Gene Phylogeny tsv\": (\"metat_lineage_results\", lineage_columns),\n",
    "        \"Metatranscriptome Expression\": (\"metat_expression_results\", None),\n",
    "    }\n",
    "    out = {}\n",
    "    for col, (key, cols) in col_map.items():\n",
    "        url = row.get(col, np.nan)\n",
    "        out[key] = read_tsv_from_url(str(url), cols) if pd.notnull(url) else None\n",
    "    return out\n",
    "\n",
    "def build_datastore(wide_df: pd.DataFrame) -> Dict[str, Dict[str, Optional[pd.DataFrame]]]:\n",
    "    store = {}\n",
    "    for _, row in wide_df.iterrows():\n",
    "        biosample_id = str(row[\"biosample_id\"])\n",
    "        store[biosample_id] = load_sample_row(row)\n",
    "    return store\n",
    "\n",
    "## datastore['biosample_id']['metag_ko_results']\n",
    "## datastore['biosample_id']['metat_ko_results'] etc.\n",
    "\n",
    "datastore = build_datastore(wide.head(3))\n",
    "len(datastore)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>img_ko_flag</th>\n",
       "      <th>ko_term</th>\n",
       "      <th>percent_identity</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subj_start</th>\n",
       "      <th>subj_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>align_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nmdc:wfmgan-11-fw38xr59.1_0000001_16096_17001</td>\n",
       "      <td>2574333214</td>\n",
       "      <td>KO:K11936</td>\n",
       "      <td>61.54</td>\n",
       "      <td>1</td>\n",
       "      <td>286</td>\n",
       "      <td>1</td>\n",
       "      <td>283</td>\n",
       "      <td>2.000000e-124</td>\n",
       "      <td>414.0</td>\n",
       "      <td>286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nmdc:wfmgan-11-fw38xr59.1_0000001_22302_23399</td>\n",
       "      <td>2524282950</td>\n",
       "      <td>KO:K02335</td>\n",
       "      <td>30.08</td>\n",
       "      <td>14</td>\n",
       "      <td>274</td>\n",
       "      <td>8</td>\n",
       "      <td>239</td>\n",
       "      <td>5.700000e-14</td>\n",
       "      <td>87.1</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nmdc:wfmgan-11-fw38xr59.1_0000001_25193_26839</td>\n",
       "      <td>646747065</td>\n",
       "      <td>KO:K00525</td>\n",
       "      <td>61.61</td>\n",
       "      <td>3</td>\n",
       "      <td>548</td>\n",
       "      <td>18</td>\n",
       "      <td>564</td>\n",
       "      <td>1.500000e-249</td>\n",
       "      <td>796.0</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         gene_id  img_ko_flag    ko_term  \\\n",
       "0  nmdc:wfmgan-11-fw38xr59.1_0000001_16096_17001   2574333214  KO:K11936   \n",
       "1  nmdc:wfmgan-11-fw38xr59.1_0000001_22302_23399   2524282950  KO:K02335   \n",
       "2  nmdc:wfmgan-11-fw38xr59.1_0000001_25193_26839    646747065  KO:K00525   \n",
       "\n",
       "   percent_identity  query_start  query_end  subj_start  subj_end  \\\n",
       "0             61.54            1        286           1       283   \n",
       "1             30.08           14        274           8       239   \n",
       "2             61.61            3        548          18       564   \n",
       "\n",
       "          evalue  bit_score  align_length  \n",
       "0  2.000000e-124      414.0           286  \n",
       "1   5.700000e-14       87.1           261  \n",
       "2  1.500000e-249      796.0           546  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View a snippet of the metagenome KEGG Orthology annotations file\n",
    "datastore[next(iter(datastore))]['metag_ko_results'].head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nmdc:bsm-11-042nd237\n",
      "nmdc:bsm-11-3p2mnz08\n",
      "nmdc:bsm-11-65a4xw75\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>n_metag_ko</th>\n",
       "      <th>n_metat_ko</th>\n",
       "      <th>n_shared_ko</th>\n",
       "      <th>n_metag_only</th>\n",
       "      <th>n_metat_only</th>\n",
       "      <th>jaccard</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nmdc:bsm-11-042nd237</td>\n",
       "      <td>9520</td>\n",
       "      <td>9888</td>\n",
       "      <td>8311</td>\n",
       "      <td>1209</td>\n",
       "      <td>1577</td>\n",
       "      <td>0.748941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nmdc:bsm-11-3p2mnz08</td>\n",
       "      <td>9485</td>\n",
       "      <td>8143</td>\n",
       "      <td>7244</td>\n",
       "      <td>2241</td>\n",
       "      <td>899</td>\n",
       "      <td>0.697612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nmdc:bsm-11-65a4xw75</td>\n",
       "      <td>8996</td>\n",
       "      <td>8617</td>\n",
       "      <td>7266</td>\n",
       "      <td>1730</td>\n",
       "      <td>1351</td>\n",
       "      <td>0.702233</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           biosample_id  n_metag_ko  n_metat_ko  n_shared_ko  n_metag_only  \\\n",
       "0  nmdc:bsm-11-042nd237        9520        9888         8311          1209   \n",
       "1  nmdc:bsm-11-3p2mnz08        9485        8143         7244          2241   \n",
       "2  nmdc:bsm-11-65a4xw75        8996        8617         7266          1730   \n",
       "\n",
       "   n_metat_only   jaccard  \n",
       "0          1577  0.748941  \n",
       "1           899  0.697612  \n",
       "2          1351  0.702233  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1) Shared/unique KO annotations\n",
    "def extract_ko_id(x: Optional[str]) -> Optional[str]:\n",
    "    if not isinstance(x, str):\n",
    "        return None\n",
    "    m = re.search(r\"K\\d{5}\", x.upper())\n",
    "    return m.group(0) if m else None\n",
    "def standardize_ko_df(df_ko: pd.DataFrame) -> pd.DataFrame:\n",
    "    if df_ko is None or len(df_ko) == 0:\n",
    "        return pd.DataFrame(columns=[\"gene_id\", \"ko_id\", \"ko_term\"])\n",
    "    for needed in [\"gene_id\", \"ko_term\"]:\n",
    "        if needed not in df_ko.columns:\n",
    "            raise ValueError(f\"KO file missing required column: {needed}.\")\n",
    "    out = df_ko.copy()\n",
    "    out[\"ko_id\"] = out[\"ko_term\"].map(extract_ko_id)\n",
    "    out = out.dropna(subset=[\"ko_id\"]).copy()\n",
    "    out[\"gene_id\"] = out[\"gene_id\"].astype(str)\n",
    "    return out[[\"gene_id\", \"ko_id\", \"ko_term\"]].drop_duplicates()\n",
    "\n",
    "def shared_ko_stats_per_sample(datastore):\n",
    "    rows = []\n",
    "    for biosample_id, items in datastore.items():\n",
    "        metag = standardize_ko_df(items.get(\"metag_ko_results\"))\n",
    "        metat = standardize_ko_df(items.get(\"metat_ko_results\"))\n",
    "        mg = set(metag[\"ko_id\"].dropna().unique())\n",
    "        mt = set(metat[\"ko_id\"].dropna().unique())\n",
    "        shared = mg & mt\n",
    "        rows.append({\n",
    "            \"biosample_id\": biosample_id,\n",
    "            \"n_metag_ko\": len(mg),\n",
    "            \"n_metat_ko\": len(mt),\n",
    "            \"n_shared_ko\": len(shared),\n",
    "            \"n_metag_only\": len(mg - mt),\n",
    "            \"n_metat_only\": len(mt - mg),\n",
    "            \"jaccard\": (len(shared)/len(mg|mt)) if (mg or mt) else float(\"nan\"),\n",
    "        })\n",
    "    return pd.DataFrame(rows)\n",
    "\n",
    "ko_overlap_df = shared_ko_stats_per_sample(datastore)\n",
    "#ko_overlap_df.to_csv(os.path.join(OUTPUT_DIR, \"ko_overlap_summary.csv\"), index=False)\n",
    "ko_overlap_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "alignmentgroup": "True",
         "hovertemplate": "category=n_metag_only<br>biosample_id=%{x}<br>count=%{y}<extra></extra>",
         "legendgroup": "n_metag_only",
         "marker": {
          "color": "#636efa",
          "pattern": {
           "shape": ""
          }
         },
         "name": "n_metag_only",
         "offsetgroup": "n_metag_only",
         "orientation": "v",
         "showlegend": true,
         "textposition": "auto",
         "type": "bar",
         "x": [
          "nmdc:bsm-11-042nd237",
          "nmdc:bsm-11-3p2mnz08",
          "nmdc:bsm-11-65a4xw75"
         ],
         "xaxis": "x",
         "y": [
          1209,
          2241,
          1730
         ],
         "yaxis": "y"
        },
        {
         "alignmentgroup": "True",
         "hovertemplate": "category=n_metat_only<br>biosample_id=%{x}<br>count=%{y}<extra></extra>",
         "legendgroup": "n_metat_only",
         "marker": {
          "color": "#EF553B",
          "pattern": {
           "shape": ""
          }
         },
         "name": "n_metat_only",
         "offsetgroup": "n_metat_only",
         "orientation": "v",
         "showlegend": true,
         "textposition": "auto",
         "type": "bar",
         "x": [
          "nmdc:bsm-11-042nd237",
          "nmdc:bsm-11-3p2mnz08",
          "nmdc:bsm-11-65a4xw75"
         ],
         "xaxis": "x",
         "y": [
          1577,
          899,
          1351
         ],
         "yaxis": "y"
        },
        {
         "alignmentgroup": "True",
         "hovertemplate": "category=n_shared_ko<br>biosample_id=%{x}<br>count=%{y}<extra></extra>",
         "legendgroup": "n_shared_ko",
         "marker": {
          "color": "#00cc96",
          "pattern": {
           "shape": ""
          }
         },
         "name": "n_shared_ko",
         "offsetgroup": "n_shared_ko",
         "orientation": "v",
         "showlegend": true,
         "textposition": "auto",
         "type": "bar",
         "x": [
          "nmdc:bsm-11-042nd237",
          "nmdc:bsm-11-3p2mnz08",
          "nmdc:bsm-11-65a4xw75"
         ],
         "xaxis": "x",
         "y": [
          8311,
          7244,
          7266
         ],
         "yaxis": "y"
        }
       ],
       "layout": {
        "barmode": "stack",
        "legend": {
         "title": {
          "text": "category"
         },
         "tracegroupgap": 0
        },
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "heatmapgl": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmapgl"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "Shared vs Unique KO annotations per biosample"
        },
        "xaxis": {
         "anchor": "y",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "biosample_id"
         }
        },
        "yaxis": {
         "anchor": "x",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "count"
         }
        }
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot: stacked bar chart shared vs unique\n",
    "if len(ko_overlap_df):\n",
    "    df_long = ko_overlap_df.melt(id_vars=[\"biosample_id\"], value_vars=[\"n_metag_only\", \"n_metat_only\", \"n_shared_ko\"],\n",
    "                                 var_name=\"category\", value_name=\"count\")\n",
    "    fig = px.bar(df_long, x=\"biosample_id\", y=\"count\", color=\"category\", barmode=\"stack\",\n",
    "                 title=\"Shared vs Unique KO annotations per biosample\")\n",
    "    #fig.write_html(os.path.join(OUTPUT_DIR, \"plot_shared_ko_barchart.html\"), include_plotlyjs=\"cdn\", full_html=True)\n",
    "fig if len(ko_overlap_df) else None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These figures indicate that a substantial portion of the metagenomic functional potential is actively expressed, while the metagenome also contains many KOs with no corresponding transcripts (possibly dormant or low-expression genes). Likewise, a number of KOs appear only in the transcriptome, which could arise from highly expressed genes of low abundance organisms or unassembled genomic fragments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ko_id</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K00844</td>\n",
       "      <td>map00521</td>\n",
       "      <td>Streptomycin biosynthesis [PATH:ko00521]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K00844</td>\n",
       "      <td>map00500</td>\n",
       "      <td>Starch and sucrose metabolism [PATH:ko00500]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K00844</td>\n",
       "      <td>map04066</td>\n",
       "      <td>HIF-1 signaling pathway [PATH:ko04066]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>K00844</td>\n",
       "      <td>map05230</td>\n",
       "      <td>Central carbon metabolism in cancer [PATH:ko05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>K00844</td>\n",
       "      <td>map00051</td>\n",
       "      <td>Fructose and mannose metabolism [PATH:ko00051]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>K00844</td>\n",
       "      <td>map00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis [PATH:ko00010]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ko_id pathway_id                                       pathway_name\n",
       "0   K00844   map00521           Streptomycin biosynthesis [PATH:ko00521]\n",
       "1   K00844   map00500       Starch and sucrose metabolism [PATH:ko00500]\n",
       "2   K00844   map04066             HIF-1 signaling pathway [PATH:ko04066]\n",
       "..     ...        ...                                                ...\n",
       "7   K00844   map05230  Central carbon metabolism in cancer [PATH:ko05...\n",
       "8   K00844   map00051     Fructose and mannose metabolism [PATH:ko00051]\n",
       "9   K00844   map00010        Glycolysis / Gluconeogenesis [PATH:ko00010]\n",
       "\n",
       "[10 rows x 3 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# 2) KEGG pathway mapping & coverage / expression\n",
    "RUN_KEGG_MAPPING=True\n",
    "CACHE_DIR=os.getcwd() \n",
    "def kegg_link_ko_to_pathways(ko_ids: List[str]) -> pd.DataFrame:\n",
    "    ko_ids = sorted(set([k for k in ko_ids if isinstance(k, str) and re.match(r\"K\\d{5}\", k)]))\n",
    "    if not ko_ids:\n",
    "        return pd.DataFrame(columns=[\"ko_id\", \"pathway_id\"])\n",
    "    cache_map_path = os.path.join(CACHE_DIR, \"kegg_ko_pathway_map.tsv\")\n",
    "    cached = pd.read_csv(cache_map_path, sep=\"\\t\", dtype=str) if os.path.exists(cache_map_path) else pd.DataFrame(columns=[\"ko_id\", \"pathway_id\"])\n",
    "    cached_set = set(map(tuple, cached[[\"ko_id\",\"pathway_id\"]].dropna().itertuples(index=False, name=None)))\n",
    "    missing = []\n",
    "    for k in ko_ids:\n",
    "        if not any(row[0]==k for row in cached_set): missing.append(k)\n",
    "    new_rows = []\n",
    "    if RUN_KEGG_MAPPING and missing:\n",
    "        import requests\n",
    "        for k in missing:\n",
    "            try:\n",
    "                url = f\"https://rest.kegg.jp/link/pathway/ko:{k}\"\n",
    "                resp = requests.get(url, timeout=30)\n",
    "                if resp.status_code==200:\n",
    "                    for line in resp.text.strip().splitlines():\n",
    "                        parts=line.split(\"\\t\")\n",
    "                        if len(parts)==2:\n",
    "                            ko_part = parts[0].split(\":\")[-1].strip().upper()\n",
    "                            path_part = parts[1].split(\":\")[-1].strip()\n",
    "                            new_rows.append((ko_part, path_part))\n",
    "            except Exception as e:\n",
    "                print(\"[KEGG] error\", k, e)\n",
    "        if new_rows:\n",
    "            add_df = pd.DataFrame(new_rows, columns=[\"ko_id\",\"pathway_id\"])\n",
    "            cached = pd.concat([cached, add_df], ignore_index=True).drop_duplicates()\n",
    "            cached.to_csv(cache_map_path, sep=\"\\t\", index=False)\n",
    "    return cached[cached[\"ko_id\"].isin(ko_ids)].drop_duplicates()\n",
    "\n",
    "def kegg_pathway_names(pathway_ids: List[str]) -> pd.DataFrame:\n",
    "    pathway_ids = sorted(set([p for p in pathway_ids if isinstance(p, str) and re.match(r\"map\\d{5}\", p)]))\n",
    "    if not pathway_ids:\n",
    "        return pd.DataFrame(columns=[\"pathway_id\",\"pathway_name\"])\n",
    "    cache_names_path = os.path.join(CACHE_DIR, \"kegg_pathway_names.tsv\")\n",
    "    cached = pd.read_csv(cache_names_path, sep=\"\\t\", dtype=str) if os.path.exists(cache_names_path) else pd.DataFrame(columns=[\"pathway_id\",\"pathway_name\"])\n",
    "    cached_dict = dict(zip(cached[\"pathway_id\"], cached[\"pathway_name\"]))\n",
    "    missing = [p for p in pathway_ids if p not in cached_dict]\n",
    "    new_rows = []\n",
    "    if RUN_KEGG_MAPPING and missing:\n",
    "        import requests\n",
    "        for p in missing:\n",
    "            try:\n",
    "                url = f\"https://rest.kegg.jp/list/{p}\"\n",
    "                resp = requests.get(url, timeout=30)\n",
    "                if resp.status_code==200:\n",
    "                    first = resp.text.strip().splitlines()[0]\n",
    "                    parts = first.split(\"\\t\")\n",
    "                    if len(parts)==2:\n",
    "                        new_rows.append((p, parts[1]))\n",
    "            except Exception as e:\n",
    "                print(\"[KEGG] name error\", p, e)\n",
    "        if new_rows:\n",
    "            add_df = pd.DataFrame(new_rows, columns=[\"pathway_id\",\"pathway_name\"])\n",
    "            cached = pd.concat([cached, add_df], ignore_index=True).drop_duplicates()\n",
    "            cached.to_csv(cache_names_path, sep=\"\\t\", index=False)\n",
    "    return cached[cached[\"pathway_id\"].isin(pathway_ids)].drop_duplicates()\n",
    "\n",
    "def build_ko_to_pathway_mapping(datastore):\n",
    "    all_kos = set()\n",
    "    for items in datastore.values():\n",
    "        for key in [\"metag_ko_results\",\"metat_ko_results\"]:\n",
    "            dfk = standardize_ko_df(items.get(key))\n",
    "            all_kos.update(dfk[\"ko_id\"].dropna().unique().tolist())\n",
    "    all_kos = sorted(all_kos)\n",
    "    if not all_kos: return pd.DataFrame(columns=[\"ko_id\",\"pathway_id\",\"pathway_name\"])\n",
    "    ko2path = kegg_link_ko_to_pathways(all_kos)\n",
    "    path_names = kegg_pathway_names(ko2path[\"pathway_id\"].unique().tolist()) if len(ko2path) else pd.DataFrame(columns=[\"pathway_id\",\"pathway_name\"])\n",
    "    return ko2path.merge(path_names, on=\"pathway_id\", how=\"left\")\n",
    "\n",
    "ko_path_map = build_ko_to_pathway_mapping(datastore)\n",
    "ko_path_map.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(            biosample_id pathway_id  \\\n",
       " 0   nmdc:bsm-11-042nd237   map00010   \n",
       " 1   nmdc:bsm-11-042nd237   map00020   \n",
       " 2   nmdc:bsm-11-042nd237   map00030   \n",
       " ..                   ...        ...   \n",
       " 7   nmdc:bsm-11-042nd237   map00061   \n",
       " 8   nmdc:bsm-11-042nd237   map00062   \n",
       " 9   nmdc:bsm-11-042nd237   map00071   \n",
       " \n",
       "                                    pathway_name  expr_sum  \n",
       " 0   Glycolysis / Gluconeogenesis [PATH:ko00010]  327421.0  \n",
       " 1      Citrate cycle (TCA cycle) [PATH:ko00020]  245707.0  \n",
       " 2      Pentose phosphate pathway [PATH:ko00030]  181284.0  \n",
       " ..                                          ...       ...  \n",
       " 7        Fatty acid biosynthesis [PATH:ko00061]  150937.0  \n",
       " 8          Fatty acid elongation [PATH:ko00062]    1820.0  \n",
       " 9         Fatty acid degradation [PATH:ko00071]  157872.0  \n",
       " \n",
       " [10 rows x 4 columns],\n",
       "             biosample_id pathway_id  \\\n",
       " 0   nmdc:bsm-11-042nd237   map00010   \n",
       " 1   nmdc:bsm-11-042nd237   map00020   \n",
       " 2   nmdc:bsm-11-042nd237   map00030   \n",
       " ..                   ...        ...   \n",
       " 7   nmdc:bsm-11-042nd237   map00061   \n",
       " 8   nmdc:bsm-11-042nd237   map00062   \n",
       " 9   nmdc:bsm-11-042nd237   map00071   \n",
       " \n",
       "                                    pathway_name  n_metag_kos  n_metat_kos  \\\n",
       " 0   Glycolysis / Gluconeogenesis [PATH:ko00010]           80           79   \n",
       " 1      Citrate cycle (TCA cycle) [PATH:ko00020]           54           55   \n",
       " 2      Pentose phosphate pathway [PATH:ko00030]           62           60   \n",
       " ..                                          ...          ...          ...   \n",
       " 7        Fatty acid biosynthesis [PATH:ko00061]           33           32   \n",
       " 8          Fatty acid elongation [PATH:ko00062]            9           15   \n",
       " 9         Fatty acid degradation [PATH:ko00071]           35           42   \n",
       " \n",
       "     n_shared_kos  \n",
       " 0             75  \n",
       " 1             53  \n",
       " 2             58  \n",
       " ..           ...  \n",
       " 7             32  \n",
       " 8              9  \n",
       " 9             33  \n",
       " \n",
       " [10 rows x 6 columns])"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def parse_expression_df(df_expr: pd.DataFrame) -> pd.DataFrame:\n",
    "    gene_col = \"img_gene_oid\"\n",
    "    expr_col = \"reads_cnt\"\n",
    "    OUT_gene_col = \"gene_id\"\n",
    "    OUT_expr_col = \"expr_value\"\n",
    "\n",
    "    if df_expr is None or len(df_expr) == 0:\n",
    "        return pd.DataFrame(columns=[OUT_gene_col, OUT_expr_col])\n",
    "    df = df_expr.copy()\n",
    "    \n",
    "    out = df[[gene_col, expr_col]].copy()\n",
    "    out.columns = [OUT_gene_col, OUT_expr_col]\n",
    "    out[OUT_expr_col] = pd.to_numeric(out[OUT_expr_col], errors=\"coerce\")\n",
    "    out = out.dropna(subset=[OUT_gene_col]).copy()\n",
    "    out[OUT_gene_col] = out[OUT_gene_col].astype(str)\n",
    "    out = out.groupby(OUT_gene_col, as_index=False)[OUT_expr_col].sum()\n",
    "    return out\n",
    "\n",
    "def compute_pathway_expression_coverage(datastore, ko_path_map):\n",
    "    rows_expr, rows_presence = [], []\n",
    "    for biosample_id, items in datastore.items():\n",
    "        metag = standardize_ko_df(items.get(\"metag_ko_results\"))\n",
    "        metat = standardize_ko_df(items.get(\"metat_ko_results\"))\n",
    "        expr  = parse_expression_df(items.get(\"metat_expression_results\"))\n",
    "        if len(metat) and len(expr):\n",
    "            metat_expr = metat.merge(expr, left_on=\"gene_id\", right_on=\"gene_id\", how=\"left\")\n",
    "        else:\n",
    "            metat_expr = pd.DataFrame(columns=[\"gene_id\",\"ko_id\",\"ko_term\",\"expr_value\"])\n",
    "\n",
    "        if len(ko_path_map)==0: continue\n",
    "\n",
    "        mg_kos = set(metag[\"ko_id\"].unique().tolist())\n",
    "        mt_kos = set(metat[\"ko_id\"].unique().tolist())\n",
    "        sh_kos = mg_kos & mt_kos\n",
    "\n",
    "        for path_id, sub in ko_path_map.groupby(\"pathway_id\"):\n",
    "            p_name = sub[\"pathway_name\"].dropna().iloc[0] if \"pathway_name\" in sub.columns and sub[\"pathway_name\"].notna().any() else path_id\n",
    "            path_kos = set(sub[\"ko_id\"].unique().tolist())\n",
    "            rows_presence.append({\n",
    "                \"biosample_id\": biosample_id,\n",
    "                \"pathway_id\": path_id, \"pathway_name\": p_name,\n",
    "                \"n_metag_kos\": len(path_kos & mg_kos),\n",
    "                \"n_metat_kos\": len(path_kos & mt_kos),\n",
    "                \"n_shared_kos\": len(path_kos & sh_kos),\n",
    "            })\n",
    "\n",
    "        if len(metat_expr):\n",
    "            metat_expr_by_ko = metat_expr.groupby(\"ko_id\", as_index=False)[\"expr_value\"].sum()\n",
    "            joined = metat_expr_by_ko.merge(ko_path_map[[\"ko_id\",\"pathway_id\",\"pathway_name\"]].drop_duplicates(),\n",
    "                                            on=\"ko_id\", how=\"left\")\n",
    "            grouped = joined.groupby([\"pathway_id\",\"pathway_name\"], as_index=False)[\"expr_value\"].sum()\n",
    "            for _, r in grouped.iterrows():\n",
    "                rows_expr.append({\n",
    "                    \"biosample_id\": biosample_id,\n",
    "                    \"pathway_id\": r[\"pathway_id\"],\n",
    "                    \"pathway_name\": r[\"pathway_name\"],\n",
    "                    \"expr_sum\": float(r[\"expr_value\"]),\n",
    "                })\n",
    "    return pd.DataFrame(rows_expr), pd.DataFrame(rows_presence)\n",
    "\n",
    "pathway_expr, pathway_presence = compute_pathway_expression_coverage(datastore, ko_path_map)\n",
    "OUTPUT_DIR=os.getcwd() \n",
    "pathway_expr.to_csv(os.path.join(OUTPUT_DIR, \"pathway_expression_by_sample.csv\"), index=False)\n",
    "pathway_presence.to_csv(os.path.join(OUTPUT_DIR, \"pathway_presence_by_sample.csv\"), index=False)\n",
    "(pathway_expr.head(10), pathway_presence.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "coloraxis": "coloraxis",
         "hovertemplate": "Biosample: %{x}<br>Pathway: %{y}<br>log1p(expr_sum): %{z}<extra></extra>",
         "name": "0",
         "type": "heatmap",
         "x": [
          "nmdc:bsm-11-042nd237",
          "nmdc:bsm-11-3p2mnz08",
          "nmdc:bsm-11-65a4xw75"
         ],
         "xaxis": "x",
         "y": [
          "Transporters [BR:ko02000]",
          "Ribosome [BR:ko03011]",
          "Ribosome [PATH:ko03010]",
          "Transcription machinery [BR:ko03021]",
          "Chaperones and folding catalysts [BR:ko03110]",
          "Mitochondrial biogenesis [BR:ko03029]",
          "Two-component system [PATH:ko02020]",
          "Exosome [BR:ko04147]",
          "Enzymes with EC numbers",
          "DNA repair and recombination proteins [BR:ko03400]",
          "Peptidases and inhibitors [BR:ko01002]",
          "Secretion system [BR:ko02044]",
          "Quorum sensing [PATH:ko02024]",
          "Oxidative phosphorylation [PATH:ko00190]",
          "Messenger RNA biogenesis [BR:ko03019]",
          "Transcription factors [BR:ko03000]",
          "Translation factors [BR:ko03012]",
          "Chromosome and associated proteins [BR:ko03036]",
          "Function unknown",
          "ABC transporters [PATH:ko02010]",
          "Bacterial motility proteins [BR:ko02035]",
          "RNA polymerase [PATH:ko03020]",
          "RNA degradation [PATH:ko03018]",
          "Pyruvate metabolism [PATH:ko00620]",
          "Butanoate metabolism [PATH:ko00650]",
          "Glyoxylate and dicarboxylate metabolism [PATH:ko00630]",
          "Cell cycle - Caulobacter [PATH:ko04112]",
          "Glycolysis / Gluconeogenesis [PATH:ko00010]",
          "Transcription",
          "Methane metabolism [PATH:ko00680]",
          "Longevity regulating pathway - worm [PATH:ko04212]",
          "Bacterial secretion system [PATH:ko03070]",
          "Purine metabolism [PATH:ko00230]",
          "Transfer RNA biogenesis [BR:ko03016]",
          "DNA replication proteins [BR:ko03032]",
          "Tuberculosis [PATH:ko05152]",
          "Other carbon fixation pathways [PATH:ko00720]",
          "Ribosome biogenesis [BR:ko03009]",
          "Cysteine and methionine metabolism [PATH:ko00270]",
          "Citrate cycle (TCA cycle) [PATH:ko00020]",
          "Alanine, aspartate and glutamate metabolism [PATH:ko00250]",
          "Legionellosis [PATH:ko05134]",
          "Parkinson disease [PATH:ko05012]",
          "Flagellar assembly [PATH:ko02040]",
          "Protein export [PATH:ko03060]",
          "Biofilm formation - Pseudomonas aeruginosa [PATH:ko02025]",
          "Amino acid related enzymes [BR:ko01007]",
          "Pathways of neurodegeneration - multiple diseases [PATH:ko05022]",
          "Lipid and atherosclerosis [PATH:ko05417]",
          "Two-component system [BR:ko02022]",
          "Pentose phosphate pathway [PATH:ko00030]",
          "Plant-pathogen interaction [PATH:ko04626]",
          "Biofilm formation - Vibrio cholerae [PATH:ko05111]",
          "Membrane trafficking [BR:ko04131]",
          "Chemical carcinogenesis - reactive oxygen species [PATH:ko05208]",
          "Propanoate metabolism [PATH:ko00640]",
          "Amyotrophic lateral sclerosis [PATH:ko05014]",
          "Valine, leucine and isoleucine degradation [PATH:ko00280]",
          "Lipid biosynthesis proteins [BR:ko01004]",
          "Diabetic cardiomyopathy [PATH:ko05415]",
          "Huntington disease [PATH:ko05016]",
          "Protein kinases [BR:ko01001]",
          "Glycine, serine and threonine metabolism [PATH:ko00260]",
          "Alzheimer disease [PATH:ko05010]",
          "Prion disease [PATH:ko05020]",
          "Biofilm formation - Escherichia coli [PATH:ko02026]",
          "Salmonella infection [PATH:ko05132]",
          "Pyrimidine metabolism [PATH:ko00240]",
          "Fatty acid degradation [PATH:ko00071]",
          "One carbon pool by folate [PATH:ko00670]",
          "Nitrogen metabolism [PATH:ko00910]",
          "General function prediction only",
          "Starch and sucrose metabolism [PATH:ko00500]",
          "Tryptophan metabolism [PATH:ko00380]",
          "Thermogenesis [PATH:ko04714]",
          "Fatty acid biosynthesis [PATH:ko00061]",
          "Arginine biosynthesis [PATH:ko00220]",
          "Sulfur metabolism [PATH:ko00920]",
          "Carbon fixation by Calvin cycle [PATH:ko00710]",
          "Amino sugar and nucleotide sugar metabolism [PATH:ko00520]",
          "Type I diabetes mellitus [PATH:ko04940]",
          "Benzoate degradation [PATH:ko00362]",
          "Pantothenate and CoA biosynthesis [PATH:ko00770]",
          "Arginine and proline metabolism [PATH:ko00330]",
          "Peptidoglycan biosynthesis and degradation proteins [BR:ko01011]",
          "Phenylalanine metabolism [PATH:ko00360]",
          "Terpenoid backbone biosynthesis [PATH:ko00900]",
          "Photosynthesis proteins [BR:ko00194]",
          "Longevity regulating pathway - multiple species [PATH:ko04213]",
          "Valine, leucine and isoleucine biosynthesis [PATH:ko00290]",
          "Aminoacyl-tRNA biosynthesis [PATH:ko00970]",
          "Cytoskeleton proteins [BR:ko04812]",
          "Cationic antimicrobial peptide (CAMP) resistance [PATH:ko01503]",
          "Chloroalkane and chloroalkene degradation [PATH:ko00625]",
          "Selenocompound metabolism [PATH:ko00450]",
          "Glutathione metabolism [PATH:ko00480]",
          "Fluid shear stress and atherosclerosis [PATH:ko05418]",
          "Photosynthesis [PATH:ko00195]",
          "beta-Lactam resistance [PATH:ko01501]",
          "Lipoic acid metabolism [PATH:ko00785]"
         ],
         "yaxis": "y",
         "z": [
          [
           14.420414876295089,
           13.430828478754483,
           13.485487282022902
          ],
          [
           13.937192649081624,
           12.982772520393235,
           13.328245198607297
          ],
          [
           13.936701114753998,
           12.982715028577179,
           13.32807588698886
          ],
          [
           13.764684522815173,
           12.781975083028277,
           13.061464092871542
          ],
          [
           13.670214229085605,
           12.742992825533408,
           12.814807795144262
          ],
          [
           13.661907061738093,
           12.581401518047574,
           12.82931761270569
          ],
          [
           13.523737462009063,
           12.626104475250317,
           12.867858149647105
          ],
          [
           13.524048015197808,
           12.688986330230728,
           12.716600224885342
          ],
          [
           13.638733999213889,
           12.466816366857413,
           12.752122545785545
          ],
          [
           13.461913728908291,
           12.700110133850455,
           12.654479100743496
          ],
          [
           13.546748370365004,
           12.310103777949534,
           12.732453654521455
          ],
          [
           13.429436227648042,
           12.257098404133217,
           12.747986968609311
          ],
          [
           13.174496519021444,
           12.298907035822593,
           12.168889854402458
          ],
          [
           13.159953916391586,
           12.061341627723477,
           12.265854559426835
          ],
          [
           12.997643336266211,
           12.037979384746851,
           12.179908325509494
          ],
          [
           12.964262907415787,
           11.923033113126388,
           12.306212054902833
          ],
          [
           12.846986616436364,
           11.938813922719271,
           12.307333268100507
          ],
          [
           12.88554677450277,
           12.119648368890674,
           12.087458014193931
          ],
          [
           13.086982801476827,
           11.867981098533312,
           11.887738907212215
          ],
          [
           12.907968940722368,
           12.019477971963042,
           11.86124862969104
          ],
          [
           12.684382308969417,
           11.809862081717386,
           12.153699821291474
          ],
          [
           12.696932268655287,
           11.925928756116681,
           12.018018689887816
          ],
          [
           12.759515132076242,
           11.844320548922585,
           12.008377518669551
          ],
          [
           12.828317096269954,
           11.865278081094223,
           11.909632331404493
          ],
          [
           12.644698475411744,
           12.106197060102916,
           11.768838193733025
          ],
          [
           12.90634935052389,
           11.677549676805148,
           11.874487447857348
          ],
          [
           12.670994635087776,
           11.763458476913785,
           12.001112764649593
          ],
          [
           12.699005137752025,
           11.80472903238118,
           11.797412438525468
          ],
          [
           12.463513270291015,
           11.457909460414072,
           12.112086089349479
          ],
          [
           12.527014188567684,
           11.766450780770516,
           11.705692279625723
          ],
          [
           12.535275668175771,
           11.709675758416305,
           11.74482259170745
          ],
          [
           12.697981469538895,
           11.402105629864984,
           11.787013049687472
          ],
          [
           12.63746212762427,
           11.54155179375855,
           11.678922837663757
          ],
          [
           12.679065552430211,
           11.298245992068148,
           11.689681563291302
          ],
          [
           12.36903722571093,
           11.664057306974144,
           11.486509618464794
          ],
          [
           12.362589096142175,
           11.533414131397544,
           11.56095343398081
          ],
          [
           12.427206178109518,
           11.45986218402762,
           11.487207576964598
          ],
          [
           12.534653109583878,
           11.222613270110179,
           11.540657356708119
          ],
          [
           12.332189745087561,
           11.37308851978243,
           11.590793502068543
          ],
          [
           12.411899118011764,
           11.29643473448612,
           11.440860020819864
          ],
          [
           12.426075269291816,
           11.098273284652896,
           11.555457995688931
          ],
          [
           12.168225388753724,
           11.487535859536768,
           11.145160684044889
          ],
          [
           12.371858233684028,
           10.984936727325564,
           11.409152577252835
          ],
          [
           12.095520900951769,
           11.344506813345266,
           11.318903727228037
          ],
          [
           12.273628484981005,
           11.165550003636666,
           11.31481593137566
          ],
          [
           12.196703895526074,
           11.052143188066092,
           11.449516971777177
          ],
          [
           12.439332850498811,
           11.014505900161963,
           11.217896469855393
          ],
          [
           12.378932918534616,
           10.929744179012097,
           11.356447078133792
          ],
          [
           12.196471703633575,
           11.274592130815032,
           11.0916510456553
          ],
          [
           12.083673346305565,
           11.125791313546788,
           11.342113418916101
          ],
          [
           12.107825657523126,
           11.255630330725127,
           11.165606615955522
          ],
          [
           11.950709317219307,
           11.217076803185387,
           11.308873503911538
          ],
          [
           12.144064254246228,
           11.06145306690375,
           11.110655247758583
          ],
          [
           12.212159085518614,
           11.018989733872505,
           11.01402841206951
          ],
          [
           12.262958522235252,
           10.752697868727662,
           11.222974153074404
          ],
          [
           12.233113717841213,
           11.03138352469794,
           10.96699066409067
          ],
          [
           12.198176553283359,
           10.799779637903594,
           11.22867748925273
          ],
          [
           12.183811699041922,
           10.930585937441325,
           11.075706497243841
          ],
          [
           12.107991129117547,
           10.9598706069865,
           11.08827753744529
          ],
          [
           12.200652985669446,
           10.776870783399007,
           11.157934816095596
          ],
          [
           12.183515127003423,
           10.83004539676606,
           11.119334578424354
          ],
          [
           12.045164183239535,
           10.856071640107553,
           11.185850931283975
          ],
          [
           12.021175775461838,
           11.005942448921688,
           11.043145810640238
          ],
          [
           12.215007609587753,
           10.768948040692958,
           11.028400915623608
          ],
          [
           12.126239082651006,
           10.708891415441741,
           11.09581784507799
          ],
          [
           11.878880217982934,
           10.882039213803797,
           11.088812562489581
          ],
          [
           11.820777744211393,
           11.014077817963646,
           10.984767161075625
          ],
          [
           11.854299951618783,
           10.790205282067918,
           11.172868811997306
          ],
          [
           11.969546191322351,
           10.747272105623404,
           11.065825532084782
          ],
          [
           11.92367657750763,
           10.637464727584328,
           11.204183174786799
          ],
          [
           11.816357847513242,
           10.782616250721516,
           11.16547923373083
          ],
          [
           11.757626964847542,
           10.876631096003067,
           11.05881093278828
          ],
          [
           11.745345853064647,
           10.839972991695802,
           11.073613976078233
          ],
          [
           11.961616076854764,
           10.657141700932726,
           10.974317460191786
          ],
          [
           12.036351370711957,
           10.49457432165487,
           11.027621378434315
          ],
          [
           11.924624435452941,
           10.61678043878903,
           10.960548414046002
          ],
          [
           11.846500652050068,
           10.578801664944281,
           11.05128675205612
          ],
          [
           12.097696476245202,
           10.514176610655552,
           10.862876087313007
          ],
          [
           11.7488667745143,
           10.86744400710848,
           10.828718203979754
          ],
          [
           11.826728397873335,
           10.753937582469849,
           10.834627487260489
          ],
          [
           11.7111367749854,
           10.985394412632534,
           10.666510548007066
          ],
          [
           11.839599859111331,
           10.591848296362112,
           10.872067083118937
          ],
          [
           11.774751122876783,
           10.566072427513433,
           10.905515516688848
          ],
          [
           12.02599457024853,
           10.415352635876781,
           10.681343461023028
          ],
          [
           11.687525621664896,
           10.421864592454954,
           10.89392054169055
          ],
          [
           11.70640124810632,
           10.528677039912846,
           10.767516099747473
          ],
          [
           11.724595445890749,
           10.485479683729912,
           10.759900883354868
          ],
          [
           11.557688483443744,
           10.634050934481495,
           10.642515908352262
          ],
          [
           11.696729256045334,
           10.501994328941953,
           10.635158087766825
          ],
          [
           11.606934799041506,
           10.45435086333365,
           10.765279859073784
          ],
          [
           11.804908275973702,
           10.450683677611769,
           10.5654536156125
          ],
          [
           11.454459137360308,
           10.704995742766139,
           10.638352326447942
          ],
          [
           11.723202597777515,
           10.106510025854115,
           10.964449893752333
          ],
          [
           11.520159238032107,
           10.549726546894638,
           10.657518144952888
          ],
          [
           11.752044722406799,
           10.083598334032917,
           10.884010508643746
          ],
          [
           11.6695701749742,
           10.484445398482496,
           10.498304999022864
          ],
          [
           11.61685295433617,
           10.24643875525505,
           10.730750330040017
          ],
          [
           11.46355663340511,
           10.56936630594188,
           10.520185560932667
          ],
          [
           11.66460738563207,
           10.034559673629849,
           10.78188948386747
          ],
          [
           11.713029636351548,
           10.284728271753428,
           10.475822112879326
          ]
         ]
        }
       ],
       "layout": {
        "coloraxis": {
         "colorbar": {
          "title": {
           "text": "log1p(expr_sum)"
          }
         },
         "colorscale": [
          [
           0,
           "#0d0887"
          ],
          [
           0.1111111111111111,
           "#46039f"
          ],
          [
           0.2222222222222222,
           "#7201a8"
          ],
          [
           0.3333333333333333,
           "#9c179e"
          ],
          [
           0.4444444444444444,
           "#bd3786"
          ],
          [
           0.5555555555555556,
           "#d8576b"
          ],
          [
           0.6666666666666666,
           "#ed7953"
          ],
          [
           0.7777777777777778,
           "#fb9f3a"
          ],
          [
           0.8888888888888888,
           "#fdca26"
          ],
          [
           1,
           "#f0f921"
          ]
         ]
        },
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "heatmapgl": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmapgl"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "KEGG pathway expression (MetaT) across biosamples"
        },
        "xaxis": {
         "anchor": "y",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "Biosample"
         }
        },
        "yaxis": {
         "anchor": "x",
         "autorange": "reversed",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "Pathway"
         }
        }
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Heatmaps\n",
    "if len(pathway_expr):\n",
    "    mat = pathway_expr.copy()\n",
    "    mat[\"log_expr\"] = np.log1p(mat[\"expr_sum\"].astype(float))\n",
    "    heat = mat.pivot_table(index=\"pathway_name\", columns=\"biosample_id\", values=\"log_expr\", fill_value=0.0).sort_index()\n",
    "    # Sort pathway by sum of log_expr across all biosamples (descending)\n",
    "    pathway_order = heat.sum(axis=1).sort_values(ascending=False).index\n",
    "    heat = heat.loc[pathway_order]\n",
    "    fig = px.imshow(heat[:100], labels=dict(x=\"Biosample\", y=\"Pathway\", color=\"log1p(expr_sum)\"),aspect='auto',\n",
    "                    title=\"KEGG pathway expression (MetaT) across biosamples\")\n",
    "    fig.write_html(os.path.join(OUTPUT_DIR, \"heatmap_pathway_expression.html\"), include_plotlyjs=\"cdn\", full_html=True)\n",
    "    fig\n",
    "else:\n",
    "    None\n",
    "fig if len(pathway_expr) else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>q_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>map00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis [PATH:ko00010]</td>\n",
       "      <td>0.367879</td>\n",
       "      <td>3</td>\n",
       "      <td>0.367879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>map99999</td>\n",
       "      <td>Others</td>\n",
       "      <td>0.367879</td>\n",
       "      <td>3</td>\n",
       "      <td>0.368574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>map99998</td>\n",
       "      <td>Others</td>\n",
       "      <td>0.367879</td>\n",
       "      <td>3</td>\n",
       "      <td>0.369270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>map00270</td>\n",
       "      <td>Cysteine and methionine metabolism [PATH:ko00270]</td>\n",
       "      <td>0.367879</td>\n",
       "      <td>3</td>\n",
       "      <td>0.380047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>map00280</td>\n",
       "      <td>Valine, leucine and isoleucine degradation [PA...</td>\n",
       "      <td>0.367879</td>\n",
       "      <td>3</td>\n",
       "      <td>0.380787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>map05231</td>\n",
       "      <td>Choline metabolism in cancer [PATH:ko05231]</td>\n",
       "      <td>0.367879</td>\n",
       "      <td>3</td>\n",
       "      <td>0.381531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_id                                       pathway_name   p_value  \\\n",
       "0     map00010        Glycolysis / Gluconeogenesis [PATH:ko00010]  0.367879   \n",
       "530   map99999                                             Others  0.367879   \n",
       "529   map99998                                             Others  0.367879   \n",
       "..         ...                                                ...       ...   \n",
       "30    map00270  Cysteine and methionine metabolism [PATH:ko00270]  0.367879   \n",
       "31    map00280  Valine, leucine and isoleucine degradation [PA...  0.367879   \n",
       "496   map05231        Choline metabolism in cancer [PATH:ko05231]  0.367879   \n",
       "\n",
       "     n_samples   q_value  \n",
       "0            3  0.367879  \n",
       "530          3  0.368574  \n",
       "529          3  0.369270  \n",
       "..         ...       ...  \n",
       "30           3  0.380047  \n",
       "31           3  0.380787  \n",
       "496          3  0.381531  \n",
       "\n",
       "[20 rows x 5 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3) Statistics: Kruskal–Wallis across biosamples (per pathway)\n",
    "def kruskal_wallis_across_samples(pathway_expr: pd.DataFrame) -> pd.DataFrame:\n",
    "    if pathway_expr is None or len(pathway_expr) == 0:\n",
    "        return pd.DataFrame(columns=[\"pathway_id\",\"pathway_name\",\"p_value\",\"q_value\",\"n_samples\"])\n",
    "    results = []\n",
    "    for path_id, grp in pathway_expr.groupby(\"pathway_id\"):\n",
    "        vectors = [sub[\"expr_sum\"].astype(float).values for _, sub in grp.groupby(\"biosample_id\")]\n",
    "        if len(vectors) >= 2 and all(len(v) > 0 for v in vectors):\n",
    "            try:\n",
    "                stat, p = stats.kruskal(*vectors)\n",
    "            except Exception:\n",
    "                p = np.nan\n",
    "            results.append((path_id, grp[\"pathway_name\"].dropna().iloc[0], p, len(vectors)))\n",
    "    res_df = pd.DataFrame(results, columns=[\"pathway_id\",\"pathway_name\",\"p_value\",\"n_samples\"]).sort_values(\"p_value\")\n",
    "    m = res_df[\"p_value\"].notna().sum()\n",
    "    if m>0:\n",
    "        ranks = res_df[\"p_value\"].rank(method=\"first\")\n",
    "        res_df[\"q_value\"] = res_df[\"p_value\"] * m / ranks\n",
    "        res_df[\"q_value\"] = res_df[\"q_value\"].clip(upper=1.0)\n",
    "    else:\n",
    "        res_df[\"q_value\"] = np.nan\n",
    "    return res_df.sort_values(\"q_value\", na_position=\"last\")\n",
    "\n",
    "kw_df = kruskal_wallis_across_samples(pathway_expr)\n",
    "kw_df.to_csv(os.path.join(OUTPUT_DIR, \"stats_kruskal_pathways_across_samples.csv\"), index=False)\n",
    "kw_df.head(20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "alignmentgroup": "True",
         "box": {
          "visible": true
         },
         "hovertemplate": "biosample_id=nmdc:bsm-11-042nd237<br>pathway_name=%{x}<br>expr_sum=%{y}<extra></extra>",
         "legendgroup": "nmdc:bsm-11-042nd237",
         "marker": {
          "color": "#636efa"
         },
         "name": "nmdc:bsm-11-042nd237",
         "offsetgroup": "nmdc:bsm-11-042nd237",
         "orientation": "v",
         "points": "all",
         "scalegroup": "True",
         "showlegend": true,
         "type": "violin",
         "x": [
          "Transporters [BR:ko02000]",
          "Two-component system [PATH:ko02020]",
          "Ribosome [PATH:ko03010]",
          "Ribosome [BR:ko03011]",
          "Transcription machinery [BR:ko03021]",
          "Mitochondrial biogenesis [BR:ko03029]",
          "Chaperones and folding catalysts [BR:ko03110]",
          "DNA repair and recombination proteins [BR:ko03400]",
          "Exosome [BR:ko04147]",
          "Enzymes with EC numbers"
         ],
         "x0": " ",
         "xaxis": "x",
         "y": [
          1831076,
          746937,
          1128839,
          1129394,
          950443,
          857611,
          864765,
          702157,
          747169,
          837966
         ],
         "y0": " ",
         "yaxis": "y"
        },
        {
         "alignmentgroup": "True",
         "box": {
          "visible": true
         },
         "hovertemplate": "biosample_id=nmdc:bsm-11-3p2mnz08<br>pathway_name=%{x}<br>expr_sum=%{y}<extra></extra>",
         "legendgroup": "nmdc:bsm-11-3p2mnz08",
         "marker": {
          "color": "#EF553B"
         },
         "name": "nmdc:bsm-11-3p2mnz08",
         "offsetgroup": "nmdc:bsm-11-3p2mnz08",
         "orientation": "v",
         "points": "all",
         "scalegroup": "True",
         "showlegend": true,
         "type": "violin",
         "x": [
          "Transporters [BR:ko02000]",
          "Two-component system [PATH:ko02020]",
          "Ribosome [PATH:ko03010]",
          "Ribosome [BR:ko03011]",
          "Transcription machinery [BR:ko03021]",
          "Mitochondrial biogenesis [BR:ko03029]",
          "Chaperones and folding catalysts [BR:ko03110]",
          "DNA repair and recombination proteins [BR:ko03400]",
          "Exosome [BR:ko04147]",
          "Enzymes with EC numbers"
         ],
         "x0": " ",
         "xaxis": "x",
         "y": [
          680666,
          304401,
          434831,
          434856,
          355746,
          291093,
          342145,
          327783,
          324157,
          259578
         ],
         "y0": " ",
         "yaxis": "y"
        },
        {
         "alignmentgroup": "True",
         "box": {
          "visible": true
         },
         "hovertemplate": "biosample_id=nmdc:bsm-11-65a4xw75<br>pathway_name=%{x}<br>expr_sum=%{y}<extra></extra>",
         "legendgroup": "nmdc:bsm-11-65a4xw75",
         "marker": {
          "color": "#00cc96"
         },
         "name": "nmdc:bsm-11-65a4xw75",
         "offsetgroup": "nmdc:bsm-11-65a4xw75",
         "orientation": "v",
         "points": "all",
         "scalegroup": "True",
         "showlegend": true,
         "type": "violin",
         "x": [
          "Transporters [BR:ko02000]",
          "Two-component system [PATH:ko02020]",
          "Ribosome [PATH:ko03010]",
          "Ribosome [BR:ko03011]",
          "Transcription machinery [BR:ko03021]",
          "Mitochondrial biogenesis [BR:ko03029]",
          "Chaperones and folding catalysts [BR:ko03110]",
          "DNA repair and recombination proteins [BR:ko03400]",
          "Exosome [BR:ko04147]",
          "Enzymes with EC numbers"
         ],
         "x0": " ",
         "xaxis": "x",
         "y": [
          718906,
          387649,
          614199,
          614303,
          470458,
          372993,
          367620,
          313162,
          333233,
          345283
         ],
         "y0": " ",
         "yaxis": "y"
        }
       ],
       "layout": {
        "legend": {
         "title": {
          "text": "biosample_id"
         },
         "tracegroupgap": 0
        },
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "heatmapgl": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmapgl"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "Expression distributions for top pathways (MetaT)"
        },
        "violinmode": "group",
        "xaxis": {
         "anchor": "y",
         "domain": [
          0,
          1
         ],
         "tickangle": 45,
         "title": {
          "text": "pathway_name"
         }
        },
        "yaxis": {
         "anchor": "x",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "expr_sum"
         }
        }
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Violin/Box for top pathways\n",
    "if len(pathway_expr):\n",
    "    ranks = (pathway_expr.groupby([\"pathway_id\",\"pathway_name\"])[\"expr_sum\"]\n",
    "             .sum().sort_values(ascending=False).head(10).reset_index())\n",
    "    top_ids = set(ranks[\"pathway_id\"].tolist())\n",
    "    sub = pathway_expr[pathway_expr[\"pathway_id\"].isin(top_ids)].copy()\n",
    "    fig = px.violin(sub, x=\"pathway_name\", y=\"expr_sum\", color=\"biosample_id\", box=True,\n",
    "                    title=\"Expression distributions for top pathways (MetaT)\", points=\"all\")\n",
    "    fig.update_xaxes(tickangle=45)\n",
    "    fig.write_html(os.path.join(OUTPUT_DIR, \"violin_top_pathways.html\"), include_plotlyjs=\"cdn\", full_html=True)\n",
    "    fig\n",
    "else:\n",
    "    None\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(            biosample_id      taxon    count\n",
       " 0   nmdc:bsm-11-042nd237    Archaea    19897\n",
       " 1   nmdc:bsm-11-042nd237   Bacteria  6277682\n",
       " 2   nmdc:bsm-11-042nd237  Eukaryota    62269\n",
       " ..                   ...        ...      ...\n",
       " 7   nmdc:bsm-11-3p2mnz08    Viruses     2285\n",
       " 8   nmdc:bsm-11-65a4xw75    Archaea    25208\n",
       " 9   nmdc:bsm-11-65a4xw75   Bacteria  4854601\n",
       " \n",
       " [10 rows x 3 columns],\n",
       "             biosample_id      taxon    expr_sum\n",
       " 0   nmdc:bsm-11-042nd237    Archaea     81199.0\n",
       " 1   nmdc:bsm-11-042nd237   Bacteria  26756923.0\n",
       " 2   nmdc:bsm-11-042nd237  Eukaryota   2891915.0\n",
       " ..                   ...        ...         ...\n",
       " 7   nmdc:bsm-11-3p2mnz08    Viruses    199839.0\n",
       " 8   nmdc:bsm-11-65a4xw75    Archaea     88750.0\n",
       " 9   nmdc:bsm-11-65a4xw75   Bacteria  13250698.0\n",
       " \n",
       " [10 rows x 3 columns],\n",
       "            biosample_id  spearman_r  p_value  n_taxa\n",
       " 0  nmdc:bsm-11-042nd237         0.8      0.2       4\n",
       " 1  nmdc:bsm-11-3p2mnz08         0.8      0.2       4\n",
       " 2  nmdc:bsm-11-65a4xw75         0.8      0.2       4)"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# 6) Taxonomy comparison: MetaG lineage vs MetaT lineage + expression\n",
    "def parse_lineage_df(df_lin: pd.DataFrame) -> pd.DataFrame:\n",
    "    if df_lin is None or len(df_lin) == 0:\n",
    "        cols = [\"gene_id\", \"lineage\", \"kingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"]\n",
    "        return pd.DataFrame(columns=cols)\n",
    "    if \"gene_id\" not in df_lin.columns or \"lineage\" not in df_lin.columns:\n",
    "        raise ValueError(\"Lineage file missing gene_id or lineage column.\")\n",
    "    def split_lineage(s: str):\n",
    "        if not isinstance(s, str): return (None,)*7\n",
    "        parts = [p.strip() for p in re.split(r\"[;|]\\s*\", s) if p.strip()]\n",
    "        ranks = [None]*7\n",
    "        for i in range(min(7, len(parts))): ranks[i] = parts[i]\n",
    "        return tuple(ranks)\n",
    "    df = df_lin[[\"gene_id\", \"lineage\"]].copy()\n",
    "    split_cols = df[\"lineage\"].map(split_lineage)\n",
    "    rank_df = pd.DataFrame(split_cols.tolist(), columns=[\"kingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"])\n",
    "    out = pd.concat([df.reset_index(drop=True), rank_df], axis=1)\n",
    "    out[\"gene_id\"] = out[\"gene_id\"].astype(str)\n",
    "    return out\n",
    "\n",
    "def taxonomy_vs_expression(datastore, rank=\"genus\"):\n",
    "    mg_rows, mt_rows, st_rows = [], [], []\n",
    "    for biosample_id, items in datastore.items():\n",
    "        mg_lin = parse_lineage_df(items.get(\"metag_lineage_results\"))\n",
    "        mt_lin = parse_lineage_df(items.get(\"metat_lineage_results\"))\n",
    "        expr   = parse_expression_df(items.get(\"metat_expression_results\"))\n",
    "        if rank not in mg_lin.columns or rank not in mt_lin.columns: continue\n",
    "\n",
    "        if len(mg_lin):\n",
    "            mg_counts = (mg_lin.dropna(subset=[rank])\n",
    "                         .groupby(rank, as_index=False)[\"gene_id\"].nunique()\n",
    "                         .rename(columns={\"gene_id\":\"count\"}))\n",
    "            for _, r in mg_counts.iterrows():\n",
    "                mg_rows.append({\"biosample_id\": biosample_id, \"taxon\": r[rank], \"count\": int(r[\"count\"])})\n",
    "\n",
    "        if len(mt_lin) and len(expr):\n",
    "            mt_join = mt_lin.merge(expr, on=\"gene_id\", how=\"left\")\n",
    "            mt_sum = (mt_join.dropna(subset=[rank])\n",
    "                      .groupby(rank, as_index=False)[\"expr_value\"].sum()\n",
    "                      .rename(columns={\"expr_value\":\"expr_sum\"}))\n",
    "            for _, r in mt_sum.iterrows():\n",
    "                mt_rows.append({\"biosample_id\": biosample_id, \"taxon\": r[rank], \"expr_sum\": float(r[\"expr_sum\"])})\n",
    "\n",
    "        if len(mg_lin) and len(mt_lin) and len(expr):\n",
    "            mgc = (mg_lin.dropna(subset=[rank]).groupby(rank, as_index=False)[\"gene_id\"].nunique().rename(columns={\"gene_id\":\"count\"}))\n",
    "            mtj = mt_lin.merge(expr, on=\"gene_id\", how=\"left\")\n",
    "            mts = (mtj.dropna(subset=[rank]).groupby(rank, as_index=False)[\"expr_value\"].sum().rename(columns={\"expr_value\":\"expr_sum\"}))\n",
    "            both = mgc.merge(mts, on=rank, how=\"inner\")\n",
    "            if len(both) >= 3:\n",
    "                r, p = stats.spearmanr(both[\"count\"].astype(float), both[\"expr_sum\"].astype(float))\n",
    "                st_rows.append({\"biosample_id\": biosample_id, \"spearman_r\": float(r), \"p_value\": float(p), \"n_taxa\": int(len(both))})\n",
    "            else:\n",
    "                st_rows.append({\"biosample_id\": biosample_id, \"spearman_r\": float(\"nan\"), \"p_value\": float(\"nan\"), \"n_taxa\": int(len(both))})\n",
    "    mg_df = pd.DataFrame(mg_rows); mt_df = pd.DataFrame(mt_rows); st_df = pd.DataFrame(st_rows)\n",
    "    mg_df.to_csv(os.path.join(OUTPUT_DIR, f\"taxonomy_metaG_counts_{rank}.csv\"), index=False)\n",
    "    mt_df.to_csv(os.path.join(OUTPUT_DIR, f\"taxonomy_metaT_expr_{rank}.csv\"), index=False)\n",
    "    st_df.to_csv(os.path.join(OUTPUT_DIR, f\"taxonomy_spearman_{rank}.csv\"), index=False)\n",
    "    return mg_df, mt_df, st_df\n",
    "\n",
    "metaG_counts_kingdom, metaT_expr_kingdom, tax_corr = taxonomy_vs_expression(datastore, rank=\"kingdom\")\n",
    "(metaG_counts_kingdom.head(10), metaT_expr_kingdom.head(10), tax_corr.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "coloraxis": "coloraxis",
         "hovertemplate": "Biosample: %{x}<br>phylum: %{y}<br>MetaG gene count: %{z}<extra></extra>",
         "name": "0",
         "type": "heatmap",
         "x": [
          "nmdc:bsm-11-042nd237",
          "nmdc:bsm-11-3p2mnz08",
          "nmdc:bsm-11-65a4xw75"
         ],
         "xaxis": "x",
         "y": [
          "Bacteria",
          "Eukaryota",
          "Archaea",
          "Viruses"
         ],
         "yaxis": "y",
         "z": [
          [
           6277682,
           5518853,
           4854601
          ],
          [
           62269,
           54983,
           37040
          ],
          [
           19897,
           10813,
           25208
          ],
          [
           3131,
           2285,
           1770
          ]
         ]
        }
       ],
       "layout": {
        "coloraxis": {
         "colorbar": {
          "title": {
           "text": "MetaG gene count"
          }
         },
         "colorscale": [
          [
           0,
           "#0d0887"
          ],
          [
           0.1111111111111111,
           "#46039f"
          ],
          [
           0.2222222222222222,
           "#7201a8"
          ],
          [
           0.3333333333333333,
           "#9c179e"
          ],
          [
           0.4444444444444444,
           "#bd3786"
          ],
          [
           0.5555555555555556,
           "#d8576b"
          ],
          [
           0.6666666666666666,
           "#ed7953"
          ],
          [
           0.7777777777777778,
           "#fb9f3a"
          ],
          [
           0.8888888888888888,
           "#fdca26"
          ],
          [
           1,
           "#f0f921"
          ]
         ]
        },
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "heatmapgl": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmapgl"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "MetaG phylum counts across biosamples"
        },
        "xaxis": {
         "anchor": "y",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "Biosample"
         }
        },
        "yaxis": {
         "anchor": "x",
         "autorange": "reversed",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "phylum"
         }
        }
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "coloraxis": "coloraxis",
         "hovertemplate": "Biosample: %{x}<br>phylum: %{y}<br>log1p(MetaT expr): %{z}<extra></extra>",
         "name": "0",
         "type": "heatmap",
         "x": [
          "nmdc:bsm-11-042nd237",
          "nmdc:bsm-11-3p2mnz08",
          "nmdc:bsm-11-65a4xw75"
         ],
         "xaxis": "x",
         "y": [
          "Bacteria",
          "Eukaryota",
          "Viruses",
          "Archaea"
         ],
         "yaxis": "y",
         "z": [
          [
           17.1023038389342,
           16.19197342518996,
           16.399560863722005
          ],
          [
           14.877429816195805,
           13.069380234974856,
           13.335635970715263
          ],
          [
           12.78249778987471,
           12.205272325359404,
           12.730283288906287
          ],
          [
           11.304670526149769,
           10.643613734290067,
           11.393589974879816
          ]
         ]
        }
       ],
       "layout": {
        "coloraxis": {
         "colorbar": {
          "title": {
           "text": "log1p(MetaT expr)"
          }
         },
         "colorscale": [
          [
           0,
           "#0d0887"
          ],
          [
           0.1111111111111111,
           "#46039f"
          ],
          [
           0.2222222222222222,
           "#7201a8"
          ],
          [
           0.3333333333333333,
           "#9c179e"
          ],
          [
           0.4444444444444444,
           "#bd3786"
          ],
          [
           0.5555555555555556,
           "#d8576b"
          ],
          [
           0.6666666666666666,
           "#ed7953"
          ],
          [
           0.7777777777777778,
           "#fb9f3a"
          ],
          [
           0.8888888888888888,
           "#fdca26"
          ],
          [
           1,
           "#f0f921"
          ]
         ]
        },
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "heatmapgl": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmapgl"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "MetaT phylum expression (log1p) across biosamples"
        },
        "xaxis": {
         "anchor": "y",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "Biosample"
         }
        },
        "yaxis": {
         "anchor": "x",
         "autorange": "reversed",
         "domain": [
          0,
          1
         ],
         "title": {
          "text": "phylum"
         }
        }
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# Heatmaps for taxonomy\n",
    "def plot_taxonomy_heatmaps(metaG_counts, metaT_expr, rank=\"phylum\"):\n",
    "    mg_html, mt_html = None, None\n",
    "    if len(metaG_counts):\n",
    "        mg_mat = metaG_counts.pivot_table(index=\"taxon\", columns=\"biosample_id\", values=\"count\", fill_value=0)\n",
    "        mg_mat = mg_mat.loc[mg_mat.sum(axis=1).sort_values(ascending=False).index]\n",
    "        fig1 = px.imshow(mg_mat, labels=dict(x=\"Biosample\", y=f\"{rank}\", color=\"MetaG gene count\"), aspect=\"auto\",\n",
    "                         title=f\"MetaG {rank} counts across biosamples\")\n",
    "        fig1.write_html(os.path.join(OUTPUT_DIR, f\"heatmap_metaG_{rank}_counts.html\"), include_plotlyjs=\"cdn\", full_html=True)\n",
    "        display(fig1)\n",
    "    if len(metaT_expr):\n",
    "        tmp = metaT_expr.copy(); tmp[\"log_expr\"] = np.log1p(tmp[\"expr_sum\"].astype(float))\n",
    "        mt_mat = tmp.pivot_table(index=\"taxon\", columns=\"biosample_id\", values=\"log_expr\", fill_value=0.0)\n",
    "        mt_mat = mt_mat.loc[mt_mat.sum(axis=1).sort_values(ascending=False).index]\n",
    "        fig2 = px.imshow(mt_mat, labels=dict(x=\"Biosample\", y=f\"{rank}\", color=\"log1p(MetaT expr)\"), aspect=\"auto\",\n",
    "                         title=f\"MetaT {rank} expression (log1p) across biosamples\")\n",
    "        fig2.write_html(os.path.join(OUTPUT_DIR, f\"heatmap_metaT_{rank}_expression.html\"), include_plotlyjs=\"cdn\", full_html=True)\n",
    "        display(fig2)\n",
    "\n",
    "plot_taxonomy_heatmaps(metaG_counts_kingdom, metaT_expr_kingdom, rank=\"phylum\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nmdc-notebook",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
