{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identifying relationships between annotated omics data in NMDC\n",
    "\n",
    "In this notebook, we’ll look at how different types of omics data can be connected using popular annotation vocabularies. We’ll focus on exploring biomolecules and KEGG pathways from a group of samples that have both metagenomic and metatranscriptomic data, all available in the NMDC Data Portal. By working through these examples, you’ll see how to bring these data types together for integrated analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "## First need to install locally the nmdc_api_utilities\n",
    "%pip install nmdc_api_utilities\n",
    "%pip install Bio\n",
    "%pip install pycirclize\n",
    "%pip install dotenv\n",
    "%pip install plotly\n",
    "%pip install seaborn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up environment variables\n",
    "**You can disregard this section of code unless you have interest in testing this on the development API.**\n",
    "\n",
    "Using pythons python-dotenv package, load the environment variables from the system. This chunk of code is used in the Github CI/CD pipelines to test our development API. The variable can be used when creating `nmdc_api_utilities` objects. If you do not have the environment variable `ENV` set in a .env file, this code with default to the variable \"prod\" - which tells nmdc_api_utilities to use the production API URL."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up environment variables\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "load_dotenv()\n",
    "# load the environment variable ENV. If it does not exist, default to \"prod\"\n",
    "ENV = os.environ.get(\"ENV\", \"prod\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Patch\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import sys\n",
    "from io import StringIO\n",
    "from Bio.KEGG import REST\n",
    "from pycirclize import Circos\n",
    "import time\n",
    "import importlib.util\n",
    "import nmdc_api_utilities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identifying relationships between annotated omics data in NMDC\n",
    "\n",
    "This notebook is an example of how different omics data types may be linked via commonly used annotation vocabularies and investigated together. In this notebook we explore KEGG pathways and taxonomy lineage identified in a set of samples that have processed metagenomics and metatranscriptomics data available in the NMDC Data Portal."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Retrieve data from the NMDC database using API endpoints\n",
    "\n",
    "### Choose data to retrieve\n",
    "\n",
    "The [NMDC Data Portal](https://data.microbiomedata.org/) is a powerful resource where you can search for samples by all kinds of criteria. For this example, we’ll focus on finding samples that have both metagenomics and metatranscriptomics data.  \n",
    "To do this, just use the Data Type filters (the “upset plot” below the interactive map) to quickly spot the right samples. In our case, these filters lead us to samples from the study [“Jeff Blanchard’s Harvard forest soil project”](https://data.microbiomedata.org/details/study/nmdc:sty-11-8ws97026).\n",
    "\n",
    "### Retrieve and filter data for Harvard forest soil study\n",
    "\n",
    "The study page linked above has the NMDC study identifier in the URL: `nmdc:sty-11-8ws97026`. We will use the [nmdc_api_utilities](https://microbiomedata.github.io/nmdc_api_utilities/) package to access the data_objects/study endpoint from the [NMDC Runtime API](https://api.microbiomedata.org/docs) to retrieve all records that represent data. This includes URLs for downloading raw data files (e.g. FASTQ) as well as processed data results output by the NMDC workflows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>name</th>\n",
       "      <th>file_size_bytes</th>\n",
       "      <th>md5_checksum</th>\n",
       "      <th>data_object_type</th>\n",
       "      <th>was_generated_by</th>\n",
       "      <th>url</th>\n",
       "      <th>description</th>\n",
       "      <th>data_category</th>\n",
       "      <th>biosample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nmdc:dobj-11-0ebr0z48</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>11862.7.224581.ACGGAAC-TGTTCCG.fastq.gz</td>\n",
       "      <td>34934525960</td>\n",
       "      <td>952d3a6780ac34a03b112884ec10cdda</td>\n",
       "      <td>Metagenome Raw Reads</td>\n",
       "      <td>nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>Metagenome Raw Reads for nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>instrument_data</td>\n",
       "      <td>nmdc:bsm-11-622k6044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nmdc:dobj-11-s5chx118</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>nmdc_wfrqc-11-840m6332.1_filtered.fastq.gz</td>\n",
       "      <td>31299212828</td>\n",
       "      <td>440a9bfb671ab0fea36359c715bb5f4e</td>\n",
       "      <td>Filtered Sequencing Reads</td>\n",
       "      <td>nmdc:wfrqc-11-840m6332.1</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>Reads QC for nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-622k6044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nmdc:dobj-11-p7q8bb28</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>nmdc_wfrqc-11-840m6332.1_filterStats.txt</td>\n",
       "      <td>801</td>\n",
       "      <td>b82b37fb53861fa34a06ca201068d727</td>\n",
       "      <td>QC Statistics</td>\n",
       "      <td>nmdc:wfrqc-11-840m6332.1</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "      <td>Reads QC summary for nmdc:dgns-11-0g2bvk46</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-622k6044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3346</th>\n",
       "      <td>nmdc:dobj-11-8pbtm711</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...</td>\n",
       "      <td>12668</td>\n",
       "      <td>da7f3a98506322255816e3c2507a11e9</td>\n",
       "      <td>Configuration toml</td>\n",
       "      <td>nmdc:wfmb-11-g1vtvp02.1</td>\n",
       "      <td>https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...</td>\n",
       "      <td>CoreMS parameters used for Lipidomics workflow.</td>\n",
       "      <td>workflow_parameter_data</td>\n",
       "      <td>nmdc:bsm-11-y52p4f86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3347</th>\n",
       "      <td>nmdc:dobj-11-869nnh02</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...</td>\n",
       "      <td>853274930</td>\n",
       "      <td>946f895d2a641e9dac6cd927715df3b8</td>\n",
       "      <td>LC-MS Lipidomics Processed Data</td>\n",
       "      <td>nmdc:wfmb-11-g1vtvp02.1</td>\n",
       "      <td>https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...</td>\n",
       "      <td>CoreMS hdf5 file representing a lipidomics dat...</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-y52p4f86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348</th>\n",
       "      <td>nmdc:dobj-11-zrzcp186</td>\n",
       "      <td>nmdc:DataObject</td>\n",
       "      <td>Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...</td>\n",
       "      <td>31914586</td>\n",
       "      <td>5ec14469476f651fceaf2ec5068ad140</td>\n",
       "      <td>LC-MS Lipidomics Results</td>\n",
       "      <td>nmdc:wfmb-11-g1vtvp02.1</td>\n",
       "      <td>https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...</td>\n",
       "      <td>Lipid annotations as a result of a lipidomics ...</td>\n",
       "      <td>processed_data</td>\n",
       "      <td>nmdc:bsm-11-y52p4f86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3349 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id             type  \\\n",
       "0     nmdc:dobj-11-0ebr0z48  nmdc:DataObject   \n",
       "1     nmdc:dobj-11-s5chx118  nmdc:DataObject   \n",
       "2     nmdc:dobj-11-p7q8bb28  nmdc:DataObject   \n",
       "...                     ...              ...   \n",
       "3346  nmdc:dobj-11-8pbtm711  nmdc:DataObject   \n",
       "3347  nmdc:dobj-11-869nnh02  nmdc:DataObject   \n",
       "3348  nmdc:dobj-11-zrzcp186  nmdc:DataObject   \n",
       "\n",
       "                                                   name  file_size_bytes  \\\n",
       "0               11862.7.224581.ACGGAAC-TGTTCCG.fastq.gz      34934525960   \n",
       "1            nmdc_wfrqc-11-840m6332.1_filtered.fastq.gz      31299212828   \n",
       "2              nmdc_wfrqc-11-840m6332.1_filterStats.txt              801   \n",
       "...                                                 ...              ...   \n",
       "3346  Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...            12668   \n",
       "3347  Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...        853274930   \n",
       "3348  Blanch_Nat_Lip_H_4_AB_O_06_NEG_25Jan18_Brandi-...         31914586   \n",
       "\n",
       "                          md5_checksum                 data_object_type  \\\n",
       "0     952d3a6780ac34a03b112884ec10cdda             Metagenome Raw Reads   \n",
       "1     440a9bfb671ab0fea36359c715bb5f4e        Filtered Sequencing Reads   \n",
       "2     b82b37fb53861fa34a06ca201068d727                    QC Statistics   \n",
       "...                                ...                              ...   \n",
       "3346  da7f3a98506322255816e3c2507a11e9               Configuration toml   \n",
       "3347  946f895d2a641e9dac6cd927715df3b8  LC-MS Lipidomics Processed Data   \n",
       "3348  5ec14469476f651fceaf2ec5068ad140         LC-MS Lipidomics Results   \n",
       "\n",
       "              was_generated_by  \\\n",
       "0        nmdc:dgns-11-0g2bvk46   \n",
       "1     nmdc:wfrqc-11-840m6332.1   \n",
       "2     nmdc:wfrqc-11-840m6332.1   \n",
       "...                        ...   \n",
       "3346   nmdc:wfmb-11-g1vtvp02.1   \n",
       "3347   nmdc:wfmb-11-g1vtvp02.1   \n",
       "3348   nmdc:wfmb-11-g1vtvp02.1   \n",
       "\n",
       "                                                    url  \\\n",
       "0     https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "1     https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "2     https://data.microbiomedata.org/data/nmdc:dgns...   \n",
       "...                                                 ...   \n",
       "3346  https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...   \n",
       "3347  https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...   \n",
       "3348  https://nmdcdemo.emsl.pnnl.gov/lipidomics/blan...   \n",
       "\n",
       "                                            description  \\\n",
       "0        Metagenome Raw Reads for nmdc:dgns-11-0g2bvk46   \n",
       "1                    Reads QC for nmdc:dgns-11-0g2bvk46   \n",
       "2            Reads QC summary for nmdc:dgns-11-0g2bvk46   \n",
       "...                                                 ...   \n",
       "3346    CoreMS parameters used for Lipidomics workflow.   \n",
       "3347  CoreMS hdf5 file representing a lipidomics dat...   \n",
       "3348  Lipid annotations as a result of a lipidomics ...   \n",
       "\n",
       "                data_category          biosample_id  \n",
       "0             instrument_data  nmdc:bsm-11-622k6044  \n",
       "1              processed_data  nmdc:bsm-11-622k6044  \n",
       "2              processed_data  nmdc:bsm-11-622k6044  \n",
       "...                       ...                   ...  \n",
       "3346  workflow_parameter_data  nmdc:bsm-11-y52p4f86  \n",
       "3347           processed_data  nmdc:bsm-11-y52p4f86  \n",
       "3348           processed_data  nmdc:bsm-11-y52p4f86  \n",
       "\n",
       "[3349 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.set_option(\"display.max_rows\", 6)\n",
    "from nmdc_api_utilities.data_object_search import DataObjectSearch\n",
    "do_client = DataObjectSearch(env=ENV)\n",
    "#get all data objects associated with this study id\n",
    "data = do_client.get_data_objects_for_studies(study_id='nmdc:sty-11-8ws97026')\n",
    "data = pd.DataFrame(data)\n",
    "\n",
    "#reformat data into dataframe\n",
    "data_objects=[]\n",
    "for index, row in data.iterrows():\n",
    "    bio_id = row['biosample_id']\n",
    "    row_out = pd.json_normalize(row['data_objects'])\n",
    "    row_out['biosample_id'] = bio_id\n",
    "    data_objects.append(row_out)\n",
    "\n",
    "data_objects = pd.concat(data_objects).reset_index(drop=True)\n",
    "display(data_objects)\n",
    "\n",
    "del data, index, row, row_out, bio_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One way of further identifying a NMDC `DataObject` record is by looking at its slot `data_object_type` (https://microbiomedata.github.io/nmdc-schema/data_object_type/), which contains a value from `FileTypeEnum` (https://microbiomedata.github.io/nmdc-schema/FileTypeEnum/). \n",
    "\n",
    "We want to look at the processed data results for our three omics types of interest in this notebook. Specifically, we want the files containing KEGG Orthology, Taxonomy Lineage, and  Expression data. These expression data is available in file for metatranscriptomics data (in \"Metatranscriptome Expression\" data object) and annotation data are found together in the metagenomics and metatranscriptomics ( in 'Annotation KEGG Orthology' and 'Gene Phylogeny tsv') results files.\n",
    "\n",
    "We filter for results files with the following `data_object_type` values:\n",
    "\n",
    "| Value | Description |\n",
    "|:-----:|:-----------:|\n",
    "|Gene Phylogeny tsv|Tab-delimited file of gene phylogeny|\n",
    "|Annotation KEGG Orthology|Tab delimited file for KO annotation|\n",
    "|Metatranscriptome Expression|Read count table output|\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>data_object_type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Annotation KEGG Orthology</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Phylogeny tsv</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Metatranscriptome Expression</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              count\n",
       "data_object_type                   \n",
       "Annotation KEGG Orthology        50\n",
       "Gene Phylogeny tsv               50\n",
       "Metatranscriptome Expression     31"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to biosamples that have KO annotations, Lineage, and expression results\n",
    "data_objects = data_objects.groupby('biosample_id').filter(lambda x: all(data_type in x['data_object_type'].values for data_type in [\n",
    "        \"Annotation KEGG Orthology\",\n",
    "        \"Metatranscriptome Expression\",\n",
    "        \"Gene Phylogeny tsv\"\n",
    "    ])).reset_index(drop=True)\n",
    "\n",
    "# Filter to the desired results file types\n",
    "results_by_biosample = data_objects[data_objects['data_object_type'].isin([\n",
    "    \"Annotation KEGG Orthology\",\n",
    "    \"Metatranscriptome Expression\", \"Gene Phylogeny tsv\"\n",
    "])][['biosample_id', 'id','data_object_type', 'url']].rename(columns={'id':'output_id'}).reset_index(drop=True)\n",
    "\n",
    "# Is there one output for each data object type for each biosample?\n",
    "results_by_biosample['data_object_type'].value_counts().to_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select data has both metaT and metaG analysis result\n",
    "Notice that Each biosample is expected to have two lineage TSV files and two KO results. Therefore, the total count should be double the expression data count (31 × 2 = 62).  However, we only observed 50. This indicates that some biosamples contain only one type of data.  \n",
    "\n",
    "To address this, we need to filter out those incomplete biosamples and retain only those with both metaG and metaT data for the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  count  count\n",
      "0  both_metaT_and_metaG     19\n",
      "1            metaT_only     12\n"
     ]
    }
   ],
   "source": [
    "results_by_biosample[\"_has_metaT\"] = results_by_biosample[\"url\"].str.contains(r\"nmdc:wfmt\", case=False, na=False)\n",
    "results_by_biosample[\"_has_metaG\"] = results_by_biosample[\"url\"].str.contains(r\"nmdc:wfmg\", case=False, na=False)\n",
    "\n",
    "mask_annotation = results_by_biosample[\"data_object_type\"] == \"Annotation KEGG Orthology\"\n",
    "mask_lineage = results_by_biosample[\"data_object_type\"] == \"Gene Phylogeny tsv\"\n",
    "\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaT\"] & mask_annotation, \"data_object_type\"] = \"MetaT Annotation KEGG Orthology\"\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaG\"] & mask_annotation, \"data_object_type\"] = \"MetaG Annotation KEGG Orthology\"\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaT\"] & mask_lineage, \"data_object_type\"] = \"MetaT Gene Phylogeny tsv\"\n",
    "results_by_biosample.loc[results_by_biosample[\"_has_metaG\"] & mask_lineage, \"data_object_type\"] = \"MetaG Gene Phylogeny tsv\"\n",
    "\n",
    "# Per-biosample aggregates\n",
    "agg = results_by_biosample.groupby(results_by_biosample[\"biosample_id\"]).agg(\n",
    "    rows_per_biosample=(\"biosample_id\", \"size\"),\n",
    "    has_metaT=(\"_has_metaT\", \"any\"),\n",
    "    has_metaG=(\"_has_metaG\", \"any\"),\n",
    "    metaT_row_count=(\"_has_metaT\", \"sum\"),\n",
    "    metaG_row_count=(\"_has_metaG\", \"sum\"),\n",
    ").reset_index()\n",
    "# Classify\n",
    "def classify(row):\n",
    "    if row[\"has_metaT\"] and row[\"has_metaG\"]:\n",
    "        return \"both_metaT_and_metaG\"\n",
    "    elif row[\"has_metaT\"] and not row[\"has_metaG\"]:\n",
    "        return \"metaT_only\"\n",
    "    elif row[\"has_metaG\"] and not row[\"has_metaT\"]:\n",
    "        return \"metaG_only\"\n",
    "    else:\n",
    "        return \"neither\"\n",
    "agg[\"category\"] = agg.apply(classify, axis=1)\n",
    "\n",
    "\n",
    "\n",
    "counts = (\n",
    "    agg[\"category\"]\n",
    "    .value_counts()\n",
    "    .reset_index()\n",
    "    .rename(columns={\"index\": \"category\", \"category\": \"count\"})\n",
    ")\n",
    "\n",
    "print(counts)\n",
    "\n",
    "\n",
    "# Filter to keep only biosamples with both\n",
    "keep_ids = set(agg.loc[agg[\"category\"] == \"both_metaT_and_metaG\", \"biosample_id\"])\n",
    "drop_ids = set(agg.loc[agg[\"category\"] != \"both_metaT_and_metaG\", \"biosample_id\"])\n",
    "filtered_results_by_biosample = results_by_biosample[results_by_biosample[\"biosample_id\"].isin(keep_ids)].drop(columns=[\"_has_metaT\", \"_has_metaG\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download selected results files\n",
    " \n",
    "Now we can use the `url` slot from the filtered `DataObject` records to read in all of the files containing the annotations of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create one URL column per data type\n",
    "filtered_results_by_biosample = filtered_results_by_biosample[['biosample_id', 'data_object_type', 'url']].pivot(\n",
    "    index='biosample_id', \n",
    "    columns='data_object_type', \n",
    "    values='url'\n",
    ").reset_index()\n",
    "\n",
    "# Function to read TSV files\n",
    "def read_tsv_from_url(url:str, column_names:None):\n",
    "    response = requests.get(url)\n",
    "    if column_names is None:\n",
    "        data = pd.read_csv(StringIO(response.text), sep='\\t')\n",
    "    else:\n",
    "        data = pd.read_csv(StringIO(response.text), sep='\\t', header=None, names=column_names)\n",
    "    return data\n",
    "\n",
    "# Read KO results\n",
    "ko_columns = [\"gene_id\", \"img_ko_flag\", \"ko_term\", \"percent_identity\",\n",
    "              \"query_start\", \"query_end\", \"subj_start\", \"subj_end\",\n",
    "              \"evalue\", \"bit_score\", \"align_length\"]\n",
    "filtered_results_by_biosample['metag_ko_results'] = filtered_results_by_biosample['MetaG Annotation KEGG Orthology'].apply(\n",
    "    lambda x: read_tsv_from_url(x, ko_columns) if pd.notnull(x) else None\n",
    ")\n",
    "filtered_results_by_biosample['metat_ko_results'] = filtered_results_by_biosample['MetaT Annotation KEGG Orthology'].apply(\n",
    "    lambda x: read_tsv_from_url(x, ko_columns) if pd.notnull(x) else None\n",
    ")\n",
    "\n",
    "# Read lineage results\n",
    "lineage_columns = [\"gene_id\", \"homolog_gene_oid\", \"homolog_taxon_oid\", \"percent_identity\",\"lineage\"]\n",
    "filtered_results_by_biosample['metag_lineage_results'] = filtered_results_by_biosample['MetaG Gene Phylogeny tsv'].apply(\n",
    "    lambda x: read_tsv_from_url(x, lineage_columns) if pd.notnull(x) else None\n",
    ")\n",
    "filtered_results_by_biosample['metat_lineage_results'] = filtered_results_by_biosample['MetaT Gene Phylogeny tsv'].apply(\n",
    "    lambda x: read_tsv_from_url(x, lineage_columns) if pd.notnull(x) else None\n",
    ")\n",
    "\n",
    "# Read expression results\n",
    "expression_columns= [\"img_gene_oid\", \"img_scaffold_oid\", \"locus_tag\", \"scaffold_accession\", \"strand\", \"locus_type\", \"length\", \"reads_cnt\", \"mean\", \"median\", \"stdev\", \"reads_cntA\", \"meanA\", \"medianA\", \"stdevA\"]\n",
    "filtered_results_by_biosample['metat_expression_results'] = filtered_results_by_biosample['Metatranscriptome Expression'].apply(\n",
    "    lambda x: read_tsv_from_url(x, expression_columns) if pd.notnull(x) else None\n",
    ")\n",
    "\n",
    "\n",
    "filtered_results_by_biosample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>output_id</th>\n",
       "      <th>data_object_type</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nmdc:bsm-11-8rn9bm20</td>\n",
       "      <td>nmdc:dobj-11-mp887905</td>\n",
       "      <td>MetaT Annotation KEGG Orthology</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nmdc:bsm-11-8rn9bm20</td>\n",
       "      <td>nmdc:dobj-11-hgv4z074</td>\n",
       "      <td>MetaT Gene Phylogeny tsv</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nmdc:bsm-11-8rn9bm20</td>\n",
       "      <td>nmdc:dobj-11-f282ne45</td>\n",
       "      <td>Metatranscriptome Expression</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>nmdc:bsm-11-6x067s27</td>\n",
       "      <td>nmdc:dobj-11-r46w3k75</td>\n",
       "      <td>Metatranscriptome Expression</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>nmdc:bsm-11-6x067s27</td>\n",
       "      <td>nmdc:dobj-11-mxe2b333</td>\n",
       "      <td>MetaG Annotation KEGG Orthology</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>nmdc:bsm-11-6x067s27</td>\n",
       "      <td>nmdc:dobj-11-myqzb759</td>\n",
       "      <td>MetaG Gene Phylogeny tsv</td>\n",
       "      <td>https://data.microbiomedata.org/data/nmdc:dgns...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             biosample_id              output_id  \\\n",
       "0    nmdc:bsm-11-8rn9bm20  nmdc:dobj-11-mp887905   \n",
       "1    nmdc:bsm-11-8rn9bm20  nmdc:dobj-11-hgv4z074   \n",
       "2    nmdc:bsm-11-8rn9bm20  nmdc:dobj-11-f282ne45   \n",
       "..                    ...                    ...   \n",
       "125  nmdc:bsm-11-6x067s27  nmdc:dobj-11-r46w3k75   \n",
       "126  nmdc:bsm-11-6x067s27  nmdc:dobj-11-mxe2b333   \n",
       "127  nmdc:bsm-11-6x067s27  nmdc:dobj-11-myqzb759   \n",
       "\n",
       "                    data_object_type  \\\n",
       "0    MetaT Annotation KEGG Orthology   \n",
       "1           MetaT Gene Phylogeny tsv   \n",
       "2       Metatranscriptome Expression   \n",
       "..                               ...   \n",
       "125     Metatranscriptome Expression   \n",
       "126  MetaG Annotation KEGG Orthology   \n",
       "127         MetaG Gene Phylogeny tsv   \n",
       "\n",
       "                                                   url  \n",
       "0    https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "1    https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "2    https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "..                                                 ...  \n",
       "125  https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "126  https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "127  https://data.microbiomedata.org/data/nmdc:dgns...  \n",
       "\n",
       "[95 rows x 4 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_results_by_biosample"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nmdc-notebook",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
