# Integrating Different Types of Omics Data

This folder includes two notebooks (in R and Python) that illustrate how metagenomic and metatranscriptomic data from the same sample can be retrieved from the [NMDC-runtime API](https://api.microbiomedata.org/docs) and integrated using KEGG annotations and taxonomy.

## R 
- [**Metagenome-Metatranscriptome Integration**](https://microbiomedata.github.io/nmdc_notebooks/metagmetat_python.html) - Static rendered notebook for viewing
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/microbiomedata/nmdc_notebooks/blob/main/metaG_metaT_integration/R/metagmetat_r.ipynb). **Running this notebook in the colab interactive environment is not recommended due to long API calls** _You need a google account to use this option_

## Python
- [**Metagenome-Metatranscriptome Integration**](https://microbiomedata.github.io/nmdc_notebooks/metagmetat_r.html) - Static rendered notebook for viewing
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/microbiomedata/nmdc_notebooks/blob/main/metaG_metaT_integration/python/metagmetat_python.ipynb). **Running this notebook in the colab interactive environment is not recommended due to long API calls** _You need a google account to use this option_
