---
title: "Untitled"
output: html_document
date: "2024-11-26"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(jsonlite)
library(KEGGREST)
library(httr)
library(circlize)
```

## Retrieve data from the NMDC database using API endpoints

### Choose data to retrieve

The NMDC data portal (https://data.microbiomedata.org/) allow us to filter data and samples according to many criteria. In this case, we use the Data Type filters (upset plot) to identify samples that have both metagenomics and metabolomics data. This returns 33 samples from the study "Riverbed sediment microbial communities from the Columbia River, Washington, USA" (https://data.microbiomedata.org/details/study/nmdc:sty-11-aygzgv51).

### Retrieve data for Columbia River sediment study

```{r}

# Retrieve all data objects associated with this study
dobj <- jsonlite::fromJSON('https://api.microbiomedata.org/data_objects/study/nmdc%3Asty-11-aygzgv51') %>% 
  unnest() %>%

  # Filter to biosamples with both metabolomics KO annotations AND metagenomics EC annotations
  group_by(biosample_id) %>%
  filter("Annotation Enzyme Commission" %in% data_object_type &
           "GC-MS Metabolomics Results" %in% data_object_type &
           "Annotation KEGG Orthology" %in% data_object_type & 
           "Protein Report" %in% data_object_type) %>%
  ungroup() %>%
  
  # Remove uninformative columns for simpler dataframe
  select(-c(alternative_identifiers, in_manifest, was_generated_by))

```

### dev - choose an example biosample to outline notebook - iterate through samples later

```{r}

# Choose an example biosample to play around with
example <- dobj %>% filter(biosample_id == "nmdc:bsm-13-peafgc08")

example_gene_ec <- read_tsv("nmdc_wfmgan-11-ddkkwt71.1_ec.tsv", col_names = FALSE, show_col_types = FALSE) # example %>% filter(data_object_type == "Annotation Enzyme Commission") %>% pull(url) %>% read_tsv(col_names = FALSE, show_col_types = FALSE)
example_gene_ko <- read_tsv("nmdc_wfmgan-11-ddkkwt71.1_ko.tsv", col_names = FALSE, show_col_types = FALSE)# example %>% filter(data_object_type == "Annotation KEGG Orthology") %>% pull(url) %>% read_tsv(col_names = FALSE, show_col_types = FALSE)
example_metab <-  example %>% filter(data_object_type == "GC-MS Metabolomics Results") %>% pull(url) %>% read_csv(show_col_types = FALSE)
example_protein <- example %>% filter(data_object_type == "Protein Report") %>% pull(url) %>% read_tsv(show_col_types = FALSE)

metag_ko_unique <- sort(unique(example_gene_ko$X3))
metag_ec_unique <- sort(unique(example_gene_ec$X3)) %>% str_subset(pattern = "-", negate = TRUE) # remove ids with dashes
metab_ko_unique <- sort(unique(example_metab$`Kegg Compound ID`))
prot_ec_unique <- sort(unique(example_protein$EC_Number)) %>% strsplit(",") %>% unlist()

names(example_gene_ec) <- c("gene_id", "img_ko_flag", "EC", "percent_identity",
                            "query_start", "query_end", "subj_start", "subj_end",
                            "evalue", "bit_score", "align_length")

names(example_gene_ko) <- c("gene_id", "img_ko_flag", "ko_term", "percent_identity",
                            "query_start", "query_end", "subj_start", "subj_end",
                            "evalue", "bit_score", "align_length")
```


### Get IDs from other KEGG databases

```{r}

# Get records for unique metabolite list
# keggGet can only return 10 at a time

# Get EC ids for each metabolite
ec_from_metabolites <- keggLink("enzyme", metab_ko_unique)

ec_from_metabolites <- data.frame(compound_id = names(ec_from_metabolites),
                                  ec_id = ec_from_metabolites) %>%
  nest(.by = compound_id, .key = "ec_id")

# Get modules for each metabolite
modules_from_metabolites <- keggLink("module", metab_ko_unique)

modules_from_metabolites <- data.frame(compound_id = names(modules_from_metabolites),
                                       module_id = modules_from_metabolites) %>%
  nest(.by = compound_id, .key = "module_id")


# Get pathways for each metabolite
pathways_from_metabolites <- keggLink("pathway", metab_ko_unique)

pathways_from_metabolites <- data.frame(compound_id = names(pathways_from_metabolites),
                                        pathway_id = pathways_from_metabolites) %>%
  nest(.by = compound_id, .key = "pathway_id")


metabolite_annotations <- data.frame(compound_id = paste0("cpd:", metab_ko_unique)) %>%
  left_join(ec_from_metabolites) %>%
  left_join(modules_from_metabolites) %>%
  left_join(pathways_from_metabolites) %>%
  mutate(compound_trimmed = substring(compound_id, 5)) %>%
  
  mutate(In_Metag_Annotations = vapply(.$ec_id, function(x) { any(x$ec_id %in% tolower(metag_ec_unique)) },
                           FUN.VALUE = TRUE)) %>%
  mutate(In_Prot_Annotations = vapply(.$ec_id, function(x) { any(x$ec_id %in% tolower(prot_ec_unique)) },
                           FUN.VALUE = TRUE))
```


```{r}

# Gather protein annotations from workflow output
prot_annotations <- example_protein %>%
  select(KO, EC_Number) %>%
  rename(ko_id = KO) %>%
  mutate(EC_Number = trimws(strsplit(EC_Number, ","))) %>%
  filter(if_any(everything(), ~ !is.na(.))) %>%
  distinct() %>%
  mutate(ko_trimmed = substring(ko_id, 4))

# Get modules for each protein
modules_from_proteins <- keggLink("module", prot_annotations$ko_trimmed)

modules_from_proteins <- data.frame(ko_id = toupper(names(modules_from_proteins)),
                                    module_id = modules_from_proteins) %>%
  nest(.by = ko_id, .key = "module_id")


# Get pathways for each protein
pathways_from_proteins <- keggLink("pathway", prot_annotations$ko_trimmed)

pathways_from_proteins <- data.frame(ko_id = toupper(names(pathways_from_proteins)),
                                     pathway_id = pathways_from_proteins) %>%
  nest(.by = ko_id, .key = "pathway_id")






# Gather gene annotations from workflow output
metag_annotations <- full_join(example_gene_ec, 
                               example_gene_ko,
                               by = join_by("gene_id")) %>%
  distinct(ko_term, EC) %>%
  filter(!is.na(EC)) %>%
  nest(.by = ko_term, .key = "EC") %>%
  rename(ko_id = ko_term, ec_id = EC) %>%
  
  mutate(ko_trimmed = substring(ko_id, 4)) %>%

  mutate(In_Metab_Annotations = vapply(.$ec_id, function(x) { any(tolower(x) %in% unique(unlist(metabolite_annotations$ec_id))) },
                           FUN.VALUE = TRUE)) %>%
  mutate(In_Prot_Annotations = ko_id %in% prot_annotations$ko_id)


# Get modules for each gene
# Break up vector of KO IDs, HTTP request URL gets too long
modules_from_genes <- split(metag_annotations$ko_trimmed, 
                            ceiling(seq_along(metag_annotations$ko_trimmed)/50)) %>%
  lapply(FUN = function(x) { keggLink("module", x) }) %>%
  list_c()


modules_from_genes <- data.frame(ko_id = toupper(names(modules_from_genes)),
                                 module_id = modules_from_genes) %>%
  nest(.by = ko_id, .key = "module_id")


# Get pathways for each gene
pathways_from_genes <- split(metag_annotations$ko_trimmed, 
                            ceiling(seq_along(metag_annotations$ko_trimmed)/50)) %>%
  lapply(FUN = function(x) { keggLink("pathway", x) }) %>%
  list_c()

pathways_from_genes <- data.frame(ko_id = toupper(names(pathways_from_genes)),
                                  pathway_id = pathways_from_genes) %>%
  nest(.by = ko_id, .key = "pathway_id")


# Add modules and pathways to gene annotations dataframe
metag_annotations <- metag_annotations %>%
  left_join(modules_from_genes) %>%
  left_join(pathways_from_genes)






# Add shared biomolecule identification counts to protein annotations dataframe
prot_annotations <- prot_annotations %>% 
  mutate(In_Metab_Annotations = vapply(.$EC_Number, function(x) { any(tolower(x) %in% unique(unlist(metabolite_annotations$ec_id))) },
                           FUN.VALUE = TRUE)) %>%
  mutate(In_Gene_Annotations = ko_id %in% metag_annotations$ko_id) %>%
  
  # Add modules and pathways to protein annotations dataframe
  left_join(modules_from_proteins) %>%
  left_join(pathways_from_proteins)
  

```

###### vis common biomolecule identifications

```{r circos-counts}

# generate counts for the plot
# scale

largest_sector_xlim <- max(length(metag_annotations$ko_id),
                           length(prot_annotations$ko_id),
                           length(metabolite_annotations$compound_id))

# gene sector

gene_sector_xlim <- (length(metag_annotations$ko_id))

gene_sector_prot_count <- (sum(metag_annotations$In_Prot_Annotations)) * (largest_sector_xlim / gene_sector_xlim)
gene_sector_both_count <- (nrow(filter(metag_annotations, In_Metab_Annotations & In_Prot_Annotations))) * (largest_sector_xlim / gene_sector_xlim)
gene_sector_metab_count <-(sum(metag_annotations$In_Metab_Annotations)) * (largest_sector_xlim / gene_sector_xlim)

# prot sector

prot_sector_xlim <- length(prot_annotations$ko_id)

prot_sector_metab_count <- sum(prot_annotations$In_Metab_Annotations) * (largest_sector_xlim / prot_sector_xlim)
prot_sector_both_count <- nrow(filter(prot_annotations, In_Metab_Annotations & In_Gene_Annotations)) * (largest_sector_xlim / prot_sector_xlim)
prot_sector_gene_count <-sum(prot_annotations$In_Gene_Annotations) * (largest_sector_xlim / prot_sector_xlim)


# metab sector

metab_sector_xlim <- length(metabolite_annotations$compound_id)

metab_sector_gene_count <- sum(metabolite_annotations$In_Metag_Annotations) * (metab_sector_xlim / prot_sector_xlim)
metab_sector_both_count <- nrow(filter(metabolite_annotations, In_Metag_Annotations & In_Prot_Annotations)) * (metab_sector_xlim / prot_sector_xlim)
metab_sector_prot_count <-sum(metabolite_annotations$In_Prot_Annotations) * (metab_sector_xlim / prot_sector_xlim)


gene_sector_xlim <- largest_sector_xlim
prot_sector_xlim <- largest_sector_xlim
metab_sector_xlim <- largest_sector_xlim


```

```{r circos}

# circos plots

plot_sector_names <- c("genes", "proteins", "metabolites")
 
# Create data
data = data.frame(
    sectors = c(rep("genes", gene_sector_xlim),
               rep("proteins", prot_sector_xlim),
               rep("metabolites", metab_sector_xlim)),
    x = c(seq(1:gene_sector_xlim),
          seq(1:prot_sector_xlim),
          seq(1:metab_sector_xlim)),
    y = c(seq(1:gene_sector_xlim),
          seq(1:prot_sector_xlim),
          seq(1:metab_sector_xlim))) %>%
  mutate(sectors = factor(sectors, levels = plot_sector_names))
 
# Initialize the plot.
circos.initialize(factors = data$sectors, x = data$x)

circos.labels(plot_sector_names, 
              x = c(gene_sector_xlim/2, prot_sector_xlim/2, metab_sector_xlim/2), 
              labels = c(paste0("genes (", nrow(metag_annotations), ")"),
                         paste0("proteins (", nrow(prot_annotations), ")"), 
                         paste0("metabolites (", nrow(metabolite_annotations), ")")),
              side = "outside")

 
# Build the regions of track #1
circos.trackPlotRegion(factors = data$sectors, y=data$y, bg.col = c("tomato", "gold", "skyblue2") , bg.border = NA)
 
# Add a link between a zone and another
circos.link("genes", c(0, gene_sector_prot_count), 
            "proteins", c(prot_sector_metab_count - prot_sector_both_count, 
                          prot_sector_metab_count - prot_sector_both_count + prot_sector_gene_count), 
            col = "#ff870099")


circos.link("proteins", c(0, prot_sector_metab_count), 
            "metabolites", c(metab_sector_gene_count - metab_sector_both_count,
                             metab_sector_gene_count - metab_sector_both_count + metab_sector_prot_count), 
            col = "#63ff0099")



circos.link("metabolites", c(0, metab_sector_gene_count), 
            "genes", c(gene_sector_prot_count - gene_sector_both_count,
                       gene_sector_prot_count - gene_sector_both_count + gene_sector_metab_count), 
            col = "#8b24cd99")

```

###### pathway coverage

```{r}

# Get unique list of pathways from all identified biomolecules
all_pathways <- c(metag_annotations$pathway_id, 
                  prot_annotations$pathway_id,
                  metabolite_annotations$pathway_id) %>%
  unlist(use.names = FALSE) %>%
  unique()

# Look at basic pathway maps, exclude KO highlighted pathway maps
all_pathways <- all_pathways[!str_detect(all_pathways, "path:ko")]

# Exclude "overview" pathway maps
all_pathways <- all_pathways[!str_detect(all_pathways, "path:map01")]



# Get all elements of each pathway

compounds_from_pathways <- split(all_pathways, ceiling(seq_along(all_pathways)/50)) %>%
  lapply(FUN = function(x) { keggLink("compound", x) }) %>%
  list_c()


compounds_from_pathways <- data.frame(pathway_id = names(compounds_from_pathways),
                                      compound_id = compounds_from_pathways) %>%
  nest(.by = pathway_id, .key = "compound_id")


ko_from_pathways <- split(all_pathways, ceiling(seq_along(all_pathways)/50)) %>%
  lapply(FUN = function(x) { keggLink("ko", x) }) %>%
  list_c()


ko_from_pathways <- data.frame(pathway_id = names(ko_from_pathways),
                               ko_id = ko_from_pathways) %>%
  nest(.by = pathway_id, .key = "ko_id")





# Count gene KOs
# Count compounds
# Total number of components of pathway

pathway_stats <- data.frame(pathway_id = all_pathways) %>%
  left_join(compounds_from_pathways) %>%
  left_join(ko_from_pathways) %>%
  mutate(compound_count = vapply(compound_id, 
                                 FUN = function(x) { ifelse(is.null(x), 0, nrow(x)) }, 
                                 FUN.VALUE = 1)) %>%
  mutate(ko_count =       vapply(ko_id, 
                                 FUN = function(x) { ifelse(is.null(x), 0, nrow(x)) }, 
                                 FUN.VALUE = 1))


# Count KOs detected in NMDC data (metag and metap)
all_ko_detected <- unique(c(metag_annotations$ko_id, prot_annotations$ko_id))


# Calculate coverage of pathways

pathway_stats <- pathway_stats %>%
  mutate(compound_coverage = map2_vec(compound_id, compound_count,
    .f = function(x, y) { ifelse(is.null(x), 0, 
             sum(x$compound_id %in% metabolite_annotations$compound_id)/y) }
  )) %>%
  mutate(ko_coverage = map2_vec(ko_id, ko_count,
    .f = function(x, y) { ifelse(is.null(x), 0, 
             sum(toupper(x$ko_id) %in% all_ko_detected)/y) }
  )) %>%
  
  mutate(total_components = compound_count + ko_count,
         total_coverage = (compound_coverage * compound_count + ko_coverage * ko_count)/total_components) %>%
  
  arrange(-total_coverage)



# What are the most covered pathways?
pull(head(pathway_stats, 5), pathway_id)


ggplot(pathway_stats) + 
  geom_point(aes(x = ko_coverage, y = compound_coverage, color = total_coverage)) +
  scale_color_gradient(low = "darkblue", high = "green") +
  coord_equal()


```


####### highest coverage pathways in each sample


```{r heatmap}




```

```{r scatterplot}



```





## Color pathway diagrams

```{r}

"https://www.genome.jp/pathway/map00270+1.8.1.10+1.8.4.3+C00051"

"https://www.genome.jp/module/M00338+K01758"

# color by what % of samples had this gene/protein/metabolite
# maybe just exclude 01 pathways (overviews) and go back to doing pathways

# "glycine, serine, and threonine metabolism"
# this highlights several EC numbers in green
browseURL(mark.pathway.by.objects("path:eco00260", c("eco:b0002", "eco:b0003")))

# "cysteine and methionine metabolism"
# "path:ecxxxxx" highlights almost all the EC numbers in purple ish???
# "path:ecoxxxxx" highlights some EC numbers in green
# "path:mapxxxx" highlights nothing
browseURL(mark.pathway.by.objects("path:map00270", "ec:1.8.1.10"))


# appears identical to the first one but different url
color.pathway.by.objects("path:eco00260", c("eco:b0002", "eco:b0003"), c("#ff0000", "#00ff00"), c("#ffff00","yellow"))

# "cysteine and methionine metabolism"
# no highlights
browseURL(color.pathway.by.objects("map00270", "C00051", "#00FF00", "#0000FF"))



```

```{r}
# mark.pathway.by.objects <- function(pathway.id, object.id.list)


pathway.id <- "map00270"
object.id.list <- c("1.8.1.10", "1.8.4.3", "C00051")

pathway.id <- sub("^path:", "", pathway.id)
object.id.list <- paste(object.id.list, collapse="+")
pathway.id <- sprintf("%s+%s", pathway.id, object.id.list)

url <- sprintf("https://www.kegg.jp/pathway/%s", pathway.id)

res <- GET(url)
stop_for_status(res, "GET KEGG pathway URL")
content <- content(res, type="text", encoding = "UTF-8")
lines <- strsplit(content, "\n", fixed=TRUE)[[1]]
urlLine <- grep("<img src=\"/kegg", lines, value=TRUE)
path <- strsplit(urlLine, '"', fixed=TRUE)[[1]][2]
browseURL(sprintf("https://www.kegg.jp%s", path))

```

```{r}

# color.pathway.by.objects <- function(pathway.id, object.id.list, fg.color.list, bg.color.list)

pathway.id <- "map00270"
object.id.list <- "C00022"
fg.color.list <- "#00FF00"
bg.color.list <- "#0000FF"

pathway.id <- sub("^path:", "", pathway.id)
if (!(length(object.id.list)==length(fg.color.list) &&
      length(fg.color.list) == length(bg.color.list))) {
    stop(paste("object.id.list, fg.color.list, and bg.color.list must",
        "all be the same length."))
}

# format identifier/color list as expected by server
payload <- paste(
    c("#ids", object.id.list),
    c("cols", paste(bg.color.list, fg.color.list, sep=',')),
    sep="\t",
    collapse="\n"
)

# fetch KEGG page from server, via a 302 redirect handled by httr
# transparently
# res <- POST(
#     url = "https://www.kegg.jp/kegg-bin/show_pathway",
#     body = list(
#         map = pathway.id,
#         multi_query = payload,
#         mode = 'color'
#     ),
#     encode="multipart"
# )
res <- GET("https://www.genome.jp/kegg-bin/show_pathway?map00270/c00022%09%2300ff00,0000ff")
res <- content(res, "text")

# extract image URL from page
img_matches <- regexpr(
    "(?<=<img src=\")[^\"]+",
    res,
    perl=T
)
img_url <- regmatches(res, img_matches)
if (length(img_url) < 1) {
    stop(
        "'color.pathway.by.objects()' ",
        "failed to extract KEGG image path from response."
    )
}
if (length(img_url) > 1) {
    stop(
        "'color.pathway.by.objects()' ",
        "unexpectedly matched multiple KEGG image paths in response."
    )
}
browseURL(sprintf("https://www.kegg.jp%s", img_url))





```


```{r}
library(rvest)
library(chromote)

asdf <- read_html_live("https://www.genome.jp/kegg-bin/show_pathway?map00270/c00022%09%2300ff00,0000ff")$html_elements(css = "img")

asdf <- read_html_live("https://www.genome.jp/")$html_elements(css = "img")



url <- "https://www.forbes.com/top-colleges/"
sess <- read_html_live(url)
sess |> 
  html_elements(".TopColleges2023_tableRowGroup__65RHn") |> 
  html_elements(".TopColleges2023_tableRow__BYOSU")

```

