# Integrating Different Types of Omics Data

This folder includes two notebooks (in R and Python) that illustrate how metagenomic, metaproteomic, and metabolomic data from the same sample can be retrieved from the [NMDC-runtime API](https://api.microbiomedata.org/docs) and integrated using KEGG annotations.

## R 
- [Static rendered Jupyter notebook](https://nbviewer.org/github/microbiomedata/nmdc_notebooks/blob/main/omics_types_integration/R/integration_notebook.ipynb). This is the recommended way to interact with the notebook. _Viewing only, not editable_
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/microbiomedata/nmdc_notebooks/blob/main/omics_types_integration/R/integration_notebook.ipynb). **Running this notebook in the colab interactive environment is not recommended due to long API calls** _You need a google account to use this option_

## Python
- [Static rendered Jupyter notebook](https://nbviewer.org/github/microbiomedata/nmdc_notebooks/blob/main/omics_types_integration/python/integration_notebook.ipynb). This is the recommended way to interact with the notebook. _Viewing only, not editable_
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/microbiomedata/nmdc_notebooks/blob/main/omics_types_integration/python/integration_notebook.ipynb). **Running this notebook in the colab interactive environment is not recommended due to long API calls** _You need a google account to use this option_
