# Integrating Different Types of Omics Data

This folder includes two notebooks (in R and Python) that illustrate how metagenomic, metaproteomic, and metabolomic data from the same sample can be retrieved from the [NMDC-runtime API](https://api.microbiomedata.org/docs) and integrated using KEGG annotations.

## Notebooks

### Python
- [**Omics Types Integration**](https://microbiomedata.github.io/nmdc_notebooks/integration_python.html) - Static rendered notebook for viewing
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/microbiomedata/nmdc_notebooks/blob/main/omics_types_integration/python/integration_python.ipynb) - Interactive notebook

### R
- [**Omics Types Integration**](https://microbiomedata.github.io/nmdc_notebooks/integration_r.html) - Static rendered notebook for viewing
- [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/microbiomedata/nmdc_notebooks/blob/main/omics_types_integration/R/integration_r.ipynb) - Interactive notebook

Or view all notebooks on our [**GitHub Pages site**](https://microbiomedata.github.io/nmdc_notebooks/).
